{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "743b73a5",
   "metadata": {},
   "source": [
    "# Analyze correlation in prob_escape scores for technical replicates under different conditions\n",
    "\n",
    "Tested 3 different selections conditions:\n",
    "* 6e5 virions in 600uL\n",
    "* 6e5 virions in 1200uL\n",
    "* 3e5 virions in 1200uL\n",
    "\n",
    "Included a technical replicate for lowest selection concentration in each condition. Random bottlenecking will be less of an issue at higher selection potencies, because there will be much fewer escape variants. So I wanted to analyze degree of correlation at selections around the IC99, where we expect to see more variant diversity."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ded27210",
   "metadata": {},
   "source": [
    "## Import python modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a2b4fb9f-ad03-4348-8c99-1dfda78b0dde",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-12-02T20:42:14.838432Z",
     "iopub.status.busy": "2022-12-02T20:42:14.837705Z",
     "iopub.status.idle": "2022-12-02T20:42:15.989733Z",
     "shell.execute_reply": "2022-12-02T20:42:15.988890Z"
    },
    "papermill": {
     "duration": 1.16293,
     "end_time": "2022-12-02T20:42:15.993021",
     "exception": false,
     "start_time": "2022-12-02T20:42:14.830091",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import altair as alt\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import polyclonal\n",
    "\n",
    "import pickle\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7fb7970b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "os.chdir('../../')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d646b41",
   "metadata": {},
   "source": [
    "## Set up functions for general analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "eaa4736e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# sample replicates are currently labeled as numbers 1-6, due to limitations in pipeline labeling\n",
    "# set up dict to convert these to readable labels\n",
    "sample_dict = {\n",
    "    1: '6e5vir_600uL_1',\n",
    "    2: '6e5vir_600uL_2',\n",
    "    3: '6e5vir_1200uL_1',\n",
    "    4: '6e5vir_1200uL_2',\n",
    "    5: '3e5vir_1200uL_1',\n",
    "    6: '3e5vir_1200uL_2'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "dfa2682e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define function to calculate correlations from prob_escape file\n",
    "def get_corr_df(\n",
    "    antibody,\n",
    "    replicate_conc, # list of concentrations that include technical replicates\n",
    "    sample_dict=sample_dict,\n",
    "    value_to_correlate='prob_escape'   \n",
    "):    \n",
    "    \n",
    "    replicate_prob_escapes = []\n",
    "    \n",
    "    for sample in sample_dict:\n",
    "        prob_escape = pd.read_csv(\n",
    "            f'results/prob_escape/libB_221202_1_{antibody}_{sample}_prob_escape.csv',\n",
    "            keep_default_na=False,\n",
    "            na_values='nan'\n",
    "        ).query(\"`no-antibody_count` >= no_antibody_count_threshold\")\n",
    "        assert prob_escape.notnull().all().all()\n",
    "        \n",
    "        # rough way to just get condition and replicate given current sample dict\n",
    "        prob_escape['selection_condition'] = \"_\".join(sample_dict[sample].split('_')[:2])\n",
    "        \n",
    "#         if len(replicate_conc) > 1:\n",
    "#             # include Ab conc in replicate label\n",
    "#             prob_escape['replicate'] = str(prob_escape['antibody_concentration']) + sample_dict[sample].split('_')[-1:][0]\n",
    "#         else:    \n",
    "#             prob_escape['replicate'] = sample_dict[sample].split('_')[-1:][0]\n",
    "        \n",
    "        # include Ab conc in replicate label\n",
    "        if len(replicate_conc) > 1:\n",
    "            prob_escape_list = []\n",
    "            \n",
    "            for conc in replicate_conc:             \n",
    "                prob_escape_conc = prob_escape[prob_escape['antibody_concentration'] == conc]\n",
    "                prob_escape_conc['replicate'] = str(conc) + '_' + sample_dict[sample].split('_')[-1:][0]\n",
    "                prob_escape_list.append(prob_escape_conc)\n",
    "            \n",
    "            prob_escape_filtered = pd.concat(prob_escape_list)\n",
    "        \n",
    "        # if only one Ab conc, leave replicates as single numbers\n",
    "        else:\n",
    "            prob_escape['replicate'] = sample_dict[sample].split('_')[-1:][0]\n",
    "            prob_escape_filtered = prob_escape[prob_escape['antibody_concentration'].isin(replicate_conc)]\n",
    "        \n",
    "        replicate_prob_escapes.append(prob_escape_filtered)\n",
    "    \n",
    "    replicate_df = pd.concat(replicate_prob_escapes)\n",
    "    \n",
    "    # get df of pearson correlations for technical replicates within a selection condition\n",
    "    corr = polyclonal.utils.tidy_to_corr(\n",
    "        replicate_df,\n",
    "        sample_col='replicate',\n",
    "        label_col='barcode',\n",
    "        value_col=value_to_correlate,\n",
    "        group_cols = ['selection_condition', 'antibody_concentration']\n",
    "    )\n",
    "    \n",
    "    return [corr, replicate_df]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e5c89de9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# function for generating heatmaps from corr df\n",
    "def corr_heatmap(\n",
    "    corr_df,\n",
    "    corr_col,\n",
    "    sample_cols,\n",
    "    group_col=None,\n",
    "    corr_range=(0, 1),\n",
    "    columns=3,\n",
    "    diverging_colors=None,\n",
    "#     scheme=None,\n",
    "):\n",
    "    \n",
    "    corr_chart = (\n",
    "        alt.Chart(corr_df)\n",
    "        .encode(\n",
    "            x=alt.X(sample_cols[0], title=None),\n",
    "            y=alt.Y(sample_cols[1], title=None),\n",
    "            color=alt.Color(\n",
    "                corr_col,\n",
    "                scale=(\n",
    "                    alt.Scale(domainMid=0, domain=corr_range)\n",
    "                    if diverging_colors\n",
    "                    else alt.Scale(domain=corr_range)\n",
    "                ),\n",
    "            ),\n",
    "            tooltip=[\n",
    "                alt.Tooltip(c, format=\".3f\") if corr_df[c].dtype == float else c\n",
    "                for c in corr_df.columns\n",
    "                if c not in {\"_label_1\", \"_label_2\"}\n",
    "            ],\n",
    "            facet=(\n",
    "                alt.Facet()\n",
    "                if group_col is None\n",
    "                else alt.Facet(group_col, columns=columns)\n",
    "            ),\n",
    "        )\n",
    "        .mark_rect(stroke=\"black\")\n",
    "        .properties(width=alt.Step(15), height=alt.Step(15))\n",
    "        .configure_axis(labelLimit=500)\n",
    "    )\n",
    "    \n",
    "    return corr_chart"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86a08854",
   "metadata": {},
   "source": [
    "## Analyze correlation between prob_escape scores for independent antibody selections\n",
    "These are technical replicates for selections at a single concentration, ~IC90 for the cocktail (1.37ng/uL), and ~IC99 + IC99.5 for AUSAB-13 (serum dilutions 0.0014 and 0.0021). i.e. the lowest concentrations I tend to use for selections.\n",
    "### 1:1 antibody cocktail"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e4a7dee8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-f26c04e92535441b88d0e105ce17ea47\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  var VEGA_DEBUG = (typeof VEGA_DEBUG == \"undefined\") ? {} : VEGA_DEBUG;\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-f26c04e92535441b88d0e105ce17ea47\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-f26c04e92535441b88d0e105ce17ea47\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@5.2.0?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function maybeLoadScript(lib, version) {\n",
       "      var key = `${lib.replace(\"-\", \"\")}_version`;\n",
       "      return (VEGA_DEBUG[key] == version) ?\n",
       "        Promise.resolve(paths[lib]) :\n",
       "        new Promise(function(resolve, reject) {\n",
       "          var s = document.createElement('script');\n",
       "          document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "          s.async = true;\n",
       "          s.onload = () => {\n",
       "            VEGA_DEBUG[key] = version;\n",
       "            return resolve(paths[lib]);\n",
       "          };\n",
       "          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "          s.src = paths[lib];\n",
       "        });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else {\n",
       "      maybeLoadScript(\"vega\", \"5\")\n",
       "        .then(() => maybeLoadScript(\"vega-lite\", \"5.2.0\"))\n",
       "        .then(() => maybeLoadScript(\"vega-embed\", \"6\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}, \"axis\": {\"labelLimit\": 500}}, \"data\": {\"name\": \"data-3a8d1b021864fce6ce24607877dd2903\"}, \"mark\": {\"type\": \"rect\", \"stroke\": \"black\"}, \"encoding\": {\"color\": {\"field\": \"correlation\", \"scale\": {\"domain\": [0, 1]}, \"type\": \"quantitative\"}, \"facet\": {\"columns\": 3, \"field\": \"selection_condition\", \"type\": \"nominal\"}, \"tooltip\": [{\"field\": \"selection_condition\", \"type\": \"nominal\"}, {\"field\": \"antibody_concentration\", \"format\": \".3f\", \"type\": \"quantitative\"}, {\"field\": \"replicate_1\", \"type\": \"nominal\"}, {\"field\": \"replicate_2\", \"type\": \"nominal\"}, {\"field\": \"correlation\", \"format\": \".3f\", \"type\": \"quantitative\"}], \"x\": {\"field\": \"replicate_1\", \"title\": null, \"type\": \"nominal\"}, \"y\": {\"field\": \"replicate_2\", \"title\": null, \"type\": \"nominal\"}}, \"height\": {\"step\": 15}, \"width\": {\"step\": 15}, \"$schema\": \"https://vega.github.io/schema/vega-lite/v5.2.0.json\", \"datasets\": {\"data-3a8d1b021864fce6ce24607877dd2903\": [{\"selection_condition\": \"3e5vir_1200uL\", \"antibody_concentration\": 1.37, \"replicate_1\": \"1\", \"replicate_2\": \"1\", \"correlation\": 1.0}, {\"selection_condition\": \"3e5vir_1200uL\", \"antibody_concentration\": 1.37, \"replicate_1\": \"2\", \"replicate_2\": \"1\", \"correlation\": 0.01272853203891343}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 1.37, \"replicate_1\": \"1\", \"replicate_2\": \"1\", \"correlation\": 1.0}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 1.37, \"replicate_1\": \"2\", \"replicate_2\": \"1\", \"correlation\": 0.030253567122834488}, {\"selection_condition\": \"6e5vir_600uL\", \"antibody_concentration\": 1.37, \"replicate_1\": \"1\", \"replicate_2\": \"1\", \"correlation\": 1.0}, {\"selection_condition\": \"6e5vir_600uL\", \"antibody_concentration\": 1.37, \"replicate_1\": \"2\", \"replicate_2\": \"1\", \"correlation\": 0.0892048890863237}, {\"selection_condition\": \"3e5vir_1200uL\", \"antibody_concentration\": 1.37, \"replicate_1\": \"1\", \"replicate_2\": \"2\", \"correlation\": 0.01272853203891343}, {\"selection_condition\": \"3e5vir_1200uL\", \"antibody_concentration\": 1.37, \"replicate_1\": \"2\", \"replicate_2\": \"2\", \"correlation\": 1.0}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 1.37, \"replicate_1\": \"1\", \"replicate_2\": \"2\", \"correlation\": 0.030253567122834488}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 1.37, \"replicate_1\": \"2\", \"replicate_2\": \"2\", \"correlation\": 1.0}, {\"selection_condition\": \"6e5vir_600uL\", \"antibody_concentration\": 1.37, \"replicate_1\": \"1\", \"replicate_2\": \"2\", \"correlation\": 0.0892048890863237}, {\"selection_condition\": \"6e5vir_600uL\", \"antibody_concentration\": 1.37, \"replicate_1\": \"2\", \"replicate_2\": \"2\", \"correlation\": 1.0}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.Chart(...)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cocktail_corr = get_corr_df(\n",
    "    antibody='1C04-5G04', \n",
    "    replicate_conc=[1.37]\n",
    ")[0]\n",
    "\n",
    "corr_heatmap(cocktail_corr, \n",
    "             corr_col='correlation', \n",
    "             sample_cols=['replicate_1', 'replicate_2'], \n",
    "             group_col='selection_condition',\n",
    "            )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "859673b7",
   "metadata": {},
   "source": [
    "Plot on smaller scale to better resolve low correlations:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "52ffd894",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-0f234eb1a8944d7c8e5f6dc2ae1a9b38\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  var VEGA_DEBUG = (typeof VEGA_DEBUG == \"undefined\") ? {} : VEGA_DEBUG;\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-0f234eb1a8944d7c8e5f6dc2ae1a9b38\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-0f234eb1a8944d7c8e5f6dc2ae1a9b38\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@5.2.0?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function maybeLoadScript(lib, version) {\n",
       "      var key = `${lib.replace(\"-\", \"\")}_version`;\n",
       "      return (VEGA_DEBUG[key] == version) ?\n",
       "        Promise.resolve(paths[lib]) :\n",
       "        new Promise(function(resolve, reject) {\n",
       "          var s = document.createElement('script');\n",
       "          document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "          s.async = true;\n",
       "          s.onload = () => {\n",
       "            VEGA_DEBUG[key] = version;\n",
       "            return resolve(paths[lib]);\n",
       "          };\n",
       "          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "          s.src = paths[lib];\n",
       "        });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else {\n",
       "      maybeLoadScript(\"vega\", \"5\")\n",
       "        .then(() => maybeLoadScript(\"vega-lite\", \"5.2.0\"))\n",
       "        .then(() => maybeLoadScript(\"vega-embed\", \"6\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}, \"axis\": {\"labelLimit\": 500}}, \"data\": {\"name\": \"data-3a8d1b021864fce6ce24607877dd2903\"}, \"mark\": {\"type\": \"rect\", \"stroke\": \"black\"}, \"encoding\": {\"color\": {\"field\": \"correlation\", \"scale\": {\"domain\": [0, 0.1]}, \"type\": \"quantitative\"}, \"facet\": {\"columns\": 3, \"field\": \"selection_condition\", \"type\": \"nominal\"}, \"tooltip\": [{\"field\": \"selection_condition\", \"type\": \"nominal\"}, {\"field\": \"antibody_concentration\", \"format\": \".3f\", \"type\": \"quantitative\"}, {\"field\": \"replicate_1\", \"type\": \"nominal\"}, {\"field\": \"replicate_2\", \"type\": \"nominal\"}, {\"field\": \"correlation\", \"format\": \".3f\", \"type\": \"quantitative\"}], \"x\": {\"field\": \"replicate_1\", \"title\": null, \"type\": \"nominal\"}, \"y\": {\"field\": \"replicate_2\", \"title\": null, \"type\": \"nominal\"}}, \"height\": {\"step\": 15}, \"width\": {\"step\": 15}, \"$schema\": \"https://vega.github.io/schema/vega-lite/v5.2.0.json\", \"datasets\": {\"data-3a8d1b021864fce6ce24607877dd2903\": [{\"selection_condition\": \"3e5vir_1200uL\", \"antibody_concentration\": 1.37, \"replicate_1\": \"1\", \"replicate_2\": \"1\", \"correlation\": 1.0}, {\"selection_condition\": \"3e5vir_1200uL\", \"antibody_concentration\": 1.37, \"replicate_1\": \"2\", \"replicate_2\": \"1\", \"correlation\": 0.01272853203891343}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 1.37, \"replicate_1\": \"1\", \"replicate_2\": \"1\", \"correlation\": 1.0}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 1.37, \"replicate_1\": \"2\", \"replicate_2\": \"1\", \"correlation\": 0.030253567122834488}, {\"selection_condition\": \"6e5vir_600uL\", \"antibody_concentration\": 1.37, \"replicate_1\": \"1\", \"replicate_2\": \"1\", \"correlation\": 1.0}, {\"selection_condition\": \"6e5vir_600uL\", \"antibody_concentration\": 1.37, \"replicate_1\": \"2\", \"replicate_2\": \"1\", \"correlation\": 0.0892048890863237}, {\"selection_condition\": \"3e5vir_1200uL\", \"antibody_concentration\": 1.37, \"replicate_1\": \"1\", \"replicate_2\": \"2\", \"correlation\": 0.01272853203891343}, {\"selection_condition\": \"3e5vir_1200uL\", \"antibody_concentration\": 1.37, \"replicate_1\": \"2\", \"replicate_2\": \"2\", \"correlation\": 1.0}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 1.37, \"replicate_1\": \"1\", \"replicate_2\": \"2\", \"correlation\": 0.030253567122834488}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 1.37, \"replicate_1\": \"2\", \"replicate_2\": \"2\", \"correlation\": 1.0}, {\"selection_condition\": \"6e5vir_600uL\", \"antibody_concentration\": 1.37, \"replicate_1\": \"1\", \"replicate_2\": \"2\", \"correlation\": 0.0892048890863237}, {\"selection_condition\": \"6e5vir_600uL\", \"antibody_concentration\": 1.37, \"replicate_1\": \"2\", \"replicate_2\": \"2\", \"correlation\": 1.0}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.Chart(...)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "corr_heatmap(cocktail_corr, \n",
    "             corr_col='correlation', \n",
    "             sample_cols=['replicate_1', 'replicate_2'], \n",
    "             group_col='selection_condition',\n",
    "             corr_range = (0, 0.1)\n",
    "            )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8238da9a",
   "metadata": {},
   "source": [
    "### AUSAB-13 serum\n",
    "*Included technical replicates for both serum selection concentrations in all conditions*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bcf412dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-9e8f6e8edc5f456990e6df5caaa5c5d5\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  var VEGA_DEBUG = (typeof VEGA_DEBUG == \"undefined\") ? {} : VEGA_DEBUG;\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-9e8f6e8edc5f456990e6df5caaa5c5d5\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-9e8f6e8edc5f456990e6df5caaa5c5d5\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@5.2.0?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function maybeLoadScript(lib, version) {\n",
       "      var key = `${lib.replace(\"-\", \"\")}_version`;\n",
       "      return (VEGA_DEBUG[key] == version) ?\n",
       "        Promise.resolve(paths[lib]) :\n",
       "        new Promise(function(resolve, reject) {\n",
       "          var s = document.createElement('script');\n",
       "          document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "          s.async = true;\n",
       "          s.onload = () => {\n",
       "            VEGA_DEBUG[key] = version;\n",
       "            return resolve(paths[lib]);\n",
       "          };\n",
       "          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "          s.src = paths[lib];\n",
       "        });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else {\n",
       "      maybeLoadScript(\"vega\", \"5\")\n",
       "        .then(() => maybeLoadScript(\"vega-lite\", \"5.2.0\"))\n",
       "        .then(() => maybeLoadScript(\"vega-embed\", \"6\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}, \"axis\": {\"labelLimit\": 500}}, \"data\": {\"name\": \"data-b7511b3a08b032e09cfb34f0c4e3f1e5\"}, \"mark\": {\"type\": \"rect\", \"stroke\": \"black\"}, \"encoding\": {\"color\": {\"field\": \"correlation\", \"scale\": {\"domain\": [0, 1]}, \"type\": \"quantitative\"}, \"facet\": {\"columns\": 3, \"field\": \"selection_condition\", \"type\": \"nominal\"}, \"tooltip\": [{\"field\": \"selection_condition\", \"type\": \"nominal\"}, {\"field\": \"antibody_concentration\", \"format\": \".3f\", \"type\": \"quantitative\"}, {\"field\": \"replicate_1\", \"type\": \"nominal\"}, {\"field\": \"replicate_2\", \"type\": \"nominal\"}, {\"field\": \"correlation\", \"format\": \".3f\", \"type\": \"quantitative\"}], \"x\": {\"field\": \"replicate_1\", \"title\": null, \"type\": \"nominal\"}, \"y\": {\"field\": \"replicate_2\", \"title\": null, \"type\": \"nominal\"}}, \"height\": {\"step\": 15}, \"width\": {\"step\": 15}, \"$schema\": \"https://vega.github.io/schema/vega-lite/v5.2.0.json\", \"datasets\": {\"data-b7511b3a08b032e09cfb34f0c4e3f1e5\": [{\"selection_condition\": \"3e5vir_1200uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"0.0014_1\", \"replicate_2\": \"0.0014_1\", \"correlation\": 1.0}, {\"selection_condition\": \"3e5vir_1200uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"0.0014_2\", \"replicate_2\": \"0.0014_1\", \"correlation\": 0.12396955448235347}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"0.0014_1\", \"replicate_2\": \"0.0014_1\", \"correlation\": 1.0}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"0.0014_2\", \"replicate_2\": \"0.0014_1\", \"correlation\": 0.13974784056088244}, {\"selection_condition\": \"6e5vir_600uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"0.0014_1\", \"replicate_2\": \"0.0014_1\", \"correlation\": 1.0}, {\"selection_condition\": \"6e5vir_600uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"0.0014_2\", \"replicate_2\": \"0.0014_1\", \"correlation\": 0.0998396270316061}, {\"selection_condition\": \"3e5vir_1200uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"0.0014_1\", \"replicate_2\": \"0.0014_2\", \"correlation\": 0.12396955448235347}, {\"selection_condition\": \"3e5vir_1200uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"0.0014_2\", \"replicate_2\": \"0.0014_2\", \"correlation\": 1.0}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"0.0014_1\", \"replicate_2\": \"0.0014_2\", \"correlation\": 0.13974784056088244}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"0.0014_2\", \"replicate_2\": \"0.0014_2\", \"correlation\": 1.0}, {\"selection_condition\": \"6e5vir_600uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"0.0014_1\", \"replicate_2\": \"0.0014_2\", \"correlation\": 0.0998396270316061}, {\"selection_condition\": \"6e5vir_600uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"0.0014_2\", \"replicate_2\": \"0.0014_2\", \"correlation\": 1.0}, {\"selection_condition\": \"3e5vir_1200uL\", \"antibody_concentration\": 0.0021, \"replicate_1\": \"0.0021_1\", \"replicate_2\": \"0.0021_1\", \"correlation\": 1.0}, {\"selection_condition\": \"3e5vir_1200uL\", \"antibody_concentration\": 0.0021, \"replicate_1\": \"0.0021_2\", \"replicate_2\": \"0.0021_1\", \"correlation\": 0.10136781801195346}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 0.0021, \"replicate_1\": \"0.0021_1\", \"replicate_2\": \"0.0021_1\", \"correlation\": 1.0}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 0.0021, \"replicate_1\": \"0.0021_2\", \"replicate_2\": \"0.0021_1\", \"correlation\": 0.1928060709084278}, {\"selection_condition\": \"6e5vir_600uL\", \"antibody_concentration\": 0.0021, \"replicate_1\": \"0.0021_1\", \"replicate_2\": \"0.0021_1\", \"correlation\": 1.0}, {\"selection_condition\": \"3e5vir_1200uL\", \"antibody_concentration\": 0.0021, \"replicate_1\": \"0.0021_1\", \"replicate_2\": \"0.0021_2\", \"correlation\": 0.10136781801195346}, {\"selection_condition\": \"3e5vir_1200uL\", \"antibody_concentration\": 0.0021, \"replicate_1\": \"0.0021_2\", \"replicate_2\": \"0.0021_2\", \"correlation\": 1.0}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 0.0021, \"replicate_1\": \"0.0021_1\", \"replicate_2\": \"0.0021_2\", \"correlation\": 0.1928060709084278}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 0.0021, \"replicate_1\": \"0.0021_2\", \"replicate_2\": \"0.0021_2\", \"correlation\": 1.0}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.Chart(...)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "serum_df = get_corr_df('AUSAB-13', [0.0014, 0.0021])[0]\n",
    "\n",
    "corr_heatmap(serum_df,\n",
    "             corr_col='correlation', \n",
    "             sample_cols=['replicate_1', 'replicate_2'], \n",
    "             group_col='selection_condition',\n",
    "            )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "97612e76",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-6a645fb29dea4dbbae82b0fa5f0026bb\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  var VEGA_DEBUG = (typeof VEGA_DEBUG == \"undefined\") ? {} : VEGA_DEBUG;\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-6a645fb29dea4dbbae82b0fa5f0026bb\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-6a645fb29dea4dbbae82b0fa5f0026bb\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@5.2.0?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function maybeLoadScript(lib, version) {\n",
       "      var key = `${lib.replace(\"-\", \"\")}_version`;\n",
       "      return (VEGA_DEBUG[key] == version) ?\n",
       "        Promise.resolve(paths[lib]) :\n",
       "        new Promise(function(resolve, reject) {\n",
       "          var s = document.createElement('script');\n",
       "          document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "          s.async = true;\n",
       "          s.onload = () => {\n",
       "            VEGA_DEBUG[key] = version;\n",
       "            return resolve(paths[lib]);\n",
       "          };\n",
       "          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "          s.src = paths[lib];\n",
       "        });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else {\n",
       "      maybeLoadScript(\"vega\", \"5\")\n",
       "        .then(() => maybeLoadScript(\"vega-lite\", \"5.2.0\"))\n",
       "        .then(() => maybeLoadScript(\"vega-embed\", \"6\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}, \"axis\": {\"labelLimit\": 500}}, \"data\": {\"name\": \"data-b7511b3a08b032e09cfb34f0c4e3f1e5\"}, \"mark\": {\"type\": \"rect\", \"stroke\": \"black\"}, \"encoding\": {\"color\": {\"field\": \"correlation\", \"scale\": {\"domain\": [0, 0.3]}, \"type\": \"quantitative\"}, \"facet\": {\"columns\": 3, \"field\": \"selection_condition\", \"type\": \"nominal\"}, \"tooltip\": [{\"field\": \"selection_condition\", \"type\": \"nominal\"}, {\"field\": \"antibody_concentration\", \"format\": \".3f\", \"type\": \"quantitative\"}, {\"field\": \"replicate_1\", \"type\": \"nominal\"}, {\"field\": \"replicate_2\", \"type\": \"nominal\"}, {\"field\": \"correlation\", \"format\": \".3f\", \"type\": \"quantitative\"}], \"x\": {\"field\": \"replicate_1\", \"title\": null, \"type\": \"nominal\"}, \"y\": {\"field\": \"replicate_2\", \"title\": null, \"type\": \"nominal\"}}, \"height\": {\"step\": 15}, \"width\": {\"step\": 15}, \"$schema\": \"https://vega.github.io/schema/vega-lite/v5.2.0.json\", \"datasets\": {\"data-b7511b3a08b032e09cfb34f0c4e3f1e5\": [{\"selection_condition\": \"3e5vir_1200uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"0.0014_1\", \"replicate_2\": \"0.0014_1\", \"correlation\": 1.0}, {\"selection_condition\": \"3e5vir_1200uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"0.0014_2\", \"replicate_2\": \"0.0014_1\", \"correlation\": 0.12396955448235347}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"0.0014_1\", \"replicate_2\": \"0.0014_1\", \"correlation\": 1.0}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"0.0014_2\", \"replicate_2\": \"0.0014_1\", \"correlation\": 0.13974784056088244}, {\"selection_condition\": \"6e5vir_600uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"0.0014_1\", \"replicate_2\": \"0.0014_1\", \"correlation\": 1.0}, {\"selection_condition\": \"6e5vir_600uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"0.0014_2\", \"replicate_2\": \"0.0014_1\", \"correlation\": 0.0998396270316061}, {\"selection_condition\": \"3e5vir_1200uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"0.0014_1\", \"replicate_2\": \"0.0014_2\", \"correlation\": 0.12396955448235347}, {\"selection_condition\": \"3e5vir_1200uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"0.0014_2\", \"replicate_2\": \"0.0014_2\", \"correlation\": 1.0}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"0.0014_1\", \"replicate_2\": \"0.0014_2\", \"correlation\": 0.13974784056088244}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"0.0014_2\", \"replicate_2\": \"0.0014_2\", \"correlation\": 1.0}, {\"selection_condition\": \"6e5vir_600uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"0.0014_1\", \"replicate_2\": \"0.0014_2\", \"correlation\": 0.0998396270316061}, {\"selection_condition\": \"6e5vir_600uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"0.0014_2\", \"replicate_2\": \"0.0014_2\", \"correlation\": 1.0}, {\"selection_condition\": \"3e5vir_1200uL\", \"antibody_concentration\": 0.0021, \"replicate_1\": \"0.0021_1\", \"replicate_2\": \"0.0021_1\", \"correlation\": 1.0}, {\"selection_condition\": \"3e5vir_1200uL\", \"antibody_concentration\": 0.0021, \"replicate_1\": \"0.0021_2\", \"replicate_2\": \"0.0021_1\", \"correlation\": 0.10136781801195346}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 0.0021, \"replicate_1\": \"0.0021_1\", \"replicate_2\": \"0.0021_1\", \"correlation\": 1.0}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 0.0021, \"replicate_1\": \"0.0021_2\", \"replicate_2\": \"0.0021_1\", \"correlation\": 0.1928060709084278}, {\"selection_condition\": \"6e5vir_600uL\", \"antibody_concentration\": 0.0021, \"replicate_1\": \"0.0021_1\", \"replicate_2\": \"0.0021_1\", \"correlation\": 1.0}, {\"selection_condition\": \"3e5vir_1200uL\", \"antibody_concentration\": 0.0021, \"replicate_1\": \"0.0021_1\", \"replicate_2\": \"0.0021_2\", \"correlation\": 0.10136781801195346}, {\"selection_condition\": \"3e5vir_1200uL\", \"antibody_concentration\": 0.0021, \"replicate_1\": \"0.0021_2\", \"replicate_2\": \"0.0021_2\", \"correlation\": 1.0}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 0.0021, \"replicate_1\": \"0.0021_1\", \"replicate_2\": \"0.0021_2\", \"correlation\": 0.1928060709084278}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 0.0021, \"replicate_1\": \"0.0021_2\", \"replicate_2\": \"0.0021_2\", \"correlation\": 1.0}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.Chart(...)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Max is <0.3, plot on smaller scale:\n",
    "corr_heatmap(serum_df, \n",
    "             corr_col='correlation', \n",
    "             sample_cols=['replicate_1', 'replicate_2'], \n",
    "             group_col='selection_condition',\n",
    "             corr_range = (0, 0.3)\n",
    "            )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ec59b5d",
   "metadata": {},
   "source": [
    "*unfortunately, error in pooling meant that I did not sequence 6e5vir_600uL_0.0021-dilution replicate to high enough coverage. So it will be missing from these analyses.*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37c6ee46",
   "metadata": {},
   "source": [
    "### Increased primer concentration in R1 PCR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2e0860fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_dict_primers = {\n",
    "    3: '6e5vir_1200uL_1',\n",
    "    4: '6e5vir_1200uL_2',\n",
    "    7: '6e5vir_1200uL-3uM-primers_1',\n",
    "    8: '6e5vir_1200uL-3uM-primers_2'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "77cd541c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-1c51d03f2ec848e1b1c153c938c98be6\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  var VEGA_DEBUG = (typeof VEGA_DEBUG == \"undefined\") ? {} : VEGA_DEBUG;\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-1c51d03f2ec848e1b1c153c938c98be6\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-1c51d03f2ec848e1b1c153c938c98be6\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@5.2.0?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function maybeLoadScript(lib, version) {\n",
       "      var key = `${lib.replace(\"-\", \"\")}_version`;\n",
       "      return (VEGA_DEBUG[key] == version) ?\n",
       "        Promise.resolve(paths[lib]) :\n",
       "        new Promise(function(resolve, reject) {\n",
       "          var s = document.createElement('script');\n",
       "          document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "          s.async = true;\n",
       "          s.onload = () => {\n",
       "            VEGA_DEBUG[key] = version;\n",
       "            return resolve(paths[lib]);\n",
       "          };\n",
       "          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "          s.src = paths[lib];\n",
       "        });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else {\n",
       "      maybeLoadScript(\"vega\", \"5\")\n",
       "        .then(() => maybeLoadScript(\"vega-lite\", \"5.2.0\"))\n",
       "        .then(() => maybeLoadScript(\"vega-embed\", \"6\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}, \"axis\": {\"labelLimit\": 500}}, \"data\": {\"name\": \"data-29eb749355f04b713cf320af54a79ad8\"}, \"mark\": {\"type\": \"rect\", \"stroke\": \"black\"}, \"encoding\": {\"color\": {\"field\": \"correlation\", \"scale\": {\"domain\": [0, 1]}, \"type\": \"quantitative\"}, \"facet\": {\"columns\": 3, \"field\": \"selection_condition\", \"type\": \"nominal\"}, \"tooltip\": [{\"field\": \"selection_condition\", \"type\": \"nominal\"}, {\"field\": \"antibody_concentration\", \"format\": \".3f\", \"type\": \"quantitative\"}, {\"field\": \"replicate_1\", \"type\": \"nominal\"}, {\"field\": \"replicate_2\", \"type\": \"nominal\"}, {\"field\": \"correlation\", \"format\": \".3f\", \"type\": \"quantitative\"}], \"x\": {\"field\": \"replicate_1\", \"title\": null, \"type\": \"nominal\"}, \"y\": {\"field\": \"replicate_2\", \"title\": null, \"type\": \"nominal\"}}, \"height\": {\"step\": 15}, \"width\": {\"step\": 15}, \"$schema\": \"https://vega.github.io/schema/vega-lite/v5.2.0.json\", \"datasets\": {\"data-29eb749355f04b713cf320af54a79ad8\": [{\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"1\", \"replicate_2\": \"1\", \"correlation\": 1.0}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"2\", \"replicate_2\": \"1\", \"correlation\": 0.13974784056088244}, {\"selection_condition\": \"6e5vir_1200uL-3uM-primers\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"1\", \"replicate_2\": \"1\", \"correlation\": 1.0}, {\"selection_condition\": \"6e5vir_1200uL-3uM-primers\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"2\", \"replicate_2\": \"1\", \"correlation\": 0.14415204986486527}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"1\", \"replicate_2\": \"2\", \"correlation\": 0.13974784056088244}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"2\", \"replicate_2\": \"2\", \"correlation\": 1.0}, {\"selection_condition\": \"6e5vir_1200uL-3uM-primers\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"1\", \"replicate_2\": \"2\", \"correlation\": 0.14415204986486527}, {\"selection_condition\": \"6e5vir_1200uL-3uM-primers\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"2\", \"replicate_2\": \"2\", \"correlation\": 1.0}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.Chart(...)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "serum_df = get_corr_df('AUSAB-13', [0.0014], sample_dict=sample_dict_primers)[0]\n",
    "\n",
    "corr_heatmap(serum_df,\n",
    "             corr_col='correlation', \n",
    "             sample_cols=['replicate_1', 'replicate_2'], \n",
    "             group_col='selection_condition',\n",
    "            )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "0a1ad348",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-01c0a75318c048af863a299ceb937d4b\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  var VEGA_DEBUG = (typeof VEGA_DEBUG == \"undefined\") ? {} : VEGA_DEBUG;\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-01c0a75318c048af863a299ceb937d4b\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-01c0a75318c048af863a299ceb937d4b\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@5.2.0?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function maybeLoadScript(lib, version) {\n",
       "      var key = `${lib.replace(\"-\", \"\")}_version`;\n",
       "      return (VEGA_DEBUG[key] == version) ?\n",
       "        Promise.resolve(paths[lib]) :\n",
       "        new Promise(function(resolve, reject) {\n",
       "          var s = document.createElement('script');\n",
       "          document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "          s.async = true;\n",
       "          s.onload = () => {\n",
       "            VEGA_DEBUG[key] = version;\n",
       "            return resolve(paths[lib]);\n",
       "          };\n",
       "          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "          s.src = paths[lib];\n",
       "        });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else {\n",
       "      maybeLoadScript(\"vega\", \"5\")\n",
       "        .then(() => maybeLoadScript(\"vega-lite\", \"5.2.0\"))\n",
       "        .then(() => maybeLoadScript(\"vega-embed\", \"6\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}, \"axis\": {\"labelLimit\": 500}}, \"data\": {\"name\": \"data-29eb749355f04b713cf320af54a79ad8\"}, \"mark\": {\"type\": \"rect\", \"stroke\": \"black\"}, \"encoding\": {\"color\": {\"field\": \"correlation\", \"scale\": {\"domain\": [0, 0.3]}, \"type\": \"quantitative\"}, \"facet\": {\"columns\": 3, \"field\": \"selection_condition\", \"type\": \"nominal\"}, \"tooltip\": [{\"field\": \"selection_condition\", \"type\": \"nominal\"}, {\"field\": \"antibody_concentration\", \"format\": \".3f\", \"type\": \"quantitative\"}, {\"field\": \"replicate_1\", \"type\": \"nominal\"}, {\"field\": \"replicate_2\", \"type\": \"nominal\"}, {\"field\": \"correlation\", \"format\": \".3f\", \"type\": \"quantitative\"}], \"x\": {\"field\": \"replicate_1\", \"title\": null, \"type\": \"nominal\"}, \"y\": {\"field\": \"replicate_2\", \"title\": null, \"type\": \"nominal\"}}, \"height\": {\"step\": 15}, \"width\": {\"step\": 15}, \"$schema\": \"https://vega.github.io/schema/vega-lite/v5.2.0.json\", \"datasets\": {\"data-29eb749355f04b713cf320af54a79ad8\": [{\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"1\", \"replicate_2\": \"1\", \"correlation\": 1.0}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"2\", \"replicate_2\": \"1\", \"correlation\": 0.13974784056088244}, {\"selection_condition\": \"6e5vir_1200uL-3uM-primers\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"1\", \"replicate_2\": \"1\", \"correlation\": 1.0}, {\"selection_condition\": \"6e5vir_1200uL-3uM-primers\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"2\", \"replicate_2\": \"1\", \"correlation\": 0.14415204986486527}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"1\", \"replicate_2\": \"2\", \"correlation\": 0.13974784056088244}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"2\", \"replicate_2\": \"2\", \"correlation\": 1.0}, {\"selection_condition\": \"6e5vir_1200uL-3uM-primers\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"1\", \"replicate_2\": \"2\", \"correlation\": 0.14415204986486527}, {\"selection_condition\": \"6e5vir_1200uL-3uM-primers\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"2\", \"replicate_2\": \"2\", \"correlation\": 1.0}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.Chart(...)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "corr_heatmap(serum_df,\n",
    "             corr_col='correlation', \n",
    "             sample_cols=['replicate_1', 'replicate_2'], \n",
    "             group_col='selection_condition',\n",
    "             corr_range = (0, 0.3)\n",
    "            )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f177ee5",
   "metadata": {},
   "source": [
    "doesn't look like primer concentration in R1 PCR is the culprit in low replicate correlation\n",
    "\n",
    "Correlations between individual variant prob_escape scores are very low for replicate selections. Serum correlations are slightly better than cocktail correlations, but still only in the range of r=0.2. \n",
    "\n",
    "This could indicate issues with bottlenecking, i.e. prob_escape scores are highly inconsistent because their starting representation in each individual selection is random, due to bottlenecking of library diversity. But these scores are calculated based on counts in the no-antibody control, as well as the neut standard. So we need to analyze those additional factors as well before making any conclusions."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb030d1f",
   "metadata": {},
   "source": [
    "## Check correlations between variant counts, not prob_escape values\n",
    "\n",
    "If bottlenecking at selection virion # is responsible for low correlation between escape score values, we would probably expect variant counts to also be poorly correlated."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6ef9d3bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-23a375bf74f94b449417d92680921594\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  var VEGA_DEBUG = (typeof VEGA_DEBUG == \"undefined\") ? {} : VEGA_DEBUG;\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-23a375bf74f94b449417d92680921594\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-23a375bf74f94b449417d92680921594\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@5.2.0?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function maybeLoadScript(lib, version) {\n",
       "      var key = `${lib.replace(\"-\", \"\")}_version`;\n",
       "      return (VEGA_DEBUG[key] == version) ?\n",
       "        Promise.resolve(paths[lib]) :\n",
       "        new Promise(function(resolve, reject) {\n",
       "          var s = document.createElement('script');\n",
       "          document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "          s.async = true;\n",
       "          s.onload = () => {\n",
       "            VEGA_DEBUG[key] = version;\n",
       "            return resolve(paths[lib]);\n",
       "          };\n",
       "          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "          s.src = paths[lib];\n",
       "        });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else {\n",
       "      maybeLoadScript(\"vega\", \"5\")\n",
       "        .then(() => maybeLoadScript(\"vega-lite\", \"5.2.0\"))\n",
       "        .then(() => maybeLoadScript(\"vega-embed\", \"6\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}, \"axis\": {\"labelLimit\": 500}}, \"data\": {\"name\": \"data-4dd4c9ef29748c695d7efd7b2aa78ded\"}, \"mark\": {\"type\": \"rect\", \"stroke\": \"black\"}, \"encoding\": {\"color\": {\"field\": \"correlation\", \"scale\": {\"domain\": [0, 1]}, \"type\": \"quantitative\"}, \"facet\": {\"columns\": 3, \"field\": \"selection_condition\", \"type\": \"nominal\"}, \"tooltip\": [{\"field\": \"selection_condition\", \"type\": \"nominal\"}, {\"field\": \"antibody_concentration\", \"format\": \".3f\", \"type\": \"quantitative\"}, {\"field\": \"replicate_1\", \"type\": \"nominal\"}, {\"field\": \"replicate_2\", \"type\": \"nominal\"}, {\"field\": \"correlation\", \"format\": \".3f\", \"type\": \"quantitative\"}], \"x\": {\"field\": \"replicate_1\", \"title\": null, \"type\": \"nominal\"}, \"y\": {\"field\": \"replicate_2\", \"title\": null, \"type\": \"nominal\"}}, \"height\": {\"step\": 15}, \"width\": {\"step\": 15}, \"$schema\": \"https://vega.github.io/schema/vega-lite/v5.2.0.json\", \"datasets\": {\"data-4dd4c9ef29748c695d7efd7b2aa78ded\": [{\"selection_condition\": \"3e5vir_1200uL\", \"antibody_concentration\": 1.37, \"replicate_1\": \"1\", \"replicate_2\": \"1\", \"correlation\": 1.0}, {\"selection_condition\": \"3e5vir_1200uL\", \"antibody_concentration\": 1.37, \"replicate_1\": \"2\", \"replicate_2\": \"1\", \"correlation\": 0.21346106498201908}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 1.37, \"replicate_1\": \"1\", \"replicate_2\": \"1\", \"correlation\": 1.0}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 1.37, \"replicate_1\": \"2\", \"replicate_2\": \"1\", \"correlation\": 0.5792124591795249}, {\"selection_condition\": \"6e5vir_600uL\", \"antibody_concentration\": 1.37, \"replicate_1\": \"1\", \"replicate_2\": \"1\", \"correlation\": 1.0}, {\"selection_condition\": \"6e5vir_600uL\", \"antibody_concentration\": 1.37, \"replicate_1\": \"2\", \"replicate_2\": \"1\", \"correlation\": 0.9278118263101212}, {\"selection_condition\": \"3e5vir_1200uL\", \"antibody_concentration\": 1.37, \"replicate_1\": \"1\", \"replicate_2\": \"2\", \"correlation\": 0.21346106498201908}, {\"selection_condition\": \"3e5vir_1200uL\", \"antibody_concentration\": 1.37, \"replicate_1\": \"2\", \"replicate_2\": \"2\", \"correlation\": 1.0}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 1.37, \"replicate_1\": \"1\", \"replicate_2\": \"2\", \"correlation\": 0.5792124591795249}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 1.37, \"replicate_1\": \"2\", \"replicate_2\": \"2\", \"correlation\": 1.0}, {\"selection_condition\": \"6e5vir_600uL\", \"antibody_concentration\": 1.37, \"replicate_1\": \"1\", \"replicate_2\": \"2\", \"correlation\": 0.9278118263101212}, {\"selection_condition\": \"6e5vir_600uL\", \"antibody_concentration\": 1.37, \"replicate_1\": \"2\", \"replicate_2\": \"2\", \"correlation\": 1.0}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.Chart(...)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cocktail_df = get_corr_df(\n",
    "    '1C04-5G04', \n",
    "    [1.37], \n",
    "    value_to_correlate = 'antibody_count'\n",
    ")[0]\n",
    "\n",
    "corr_heatmap(cocktail_df, \n",
    "             corr_col='correlation', \n",
    "             sample_cols=['replicate_1', 'replicate_2'], \n",
    "             group_col='selection_condition',\n",
    "            )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "baf3705d",
   "metadata": {},
   "source": [
    "For antibody cocktail selections, at the lowest selection concentration (roughly IC90), variant counts are actually quite well correlated. r=0.93 for 6e5vir in 600uL, and 0.58 for 6e5 vir in 1200uL. \n",
    "\n",
    "Side note: this drop is a little surprising. Same number of virions, so we wouldn't expect large changes in correlation in a technical replicate. But if we look at avg prob escape for wildtype variants -\n",
    "* avg_prob_escape=0.12 for 6e5vir in 600uL\n",
    "* avg_prob_escape=0.02 for 6e5vir in 1200uL\n",
    "\n",
    "So this is probably affected by antibody saturation. Lots of free virions in 600uL selections - counts also reflect baseline library composition, which is going to be consistent between replicates. 1200uL selections are more representative of neutralization, and therefore more variable between replicates. We should keep this in mind, because it's tempting to say that the 3e5 drop (r=0.213) indicates further bottlenecking with reduced virion #. But we can't actually fully disentangle that from Ab saturation effects at an even lower TCID50/uL. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "619099f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-ba1d5bb329d140eb8d8ca39bbbc9132d\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  var VEGA_DEBUG = (typeof VEGA_DEBUG == \"undefined\") ? {} : VEGA_DEBUG;\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-ba1d5bb329d140eb8d8ca39bbbc9132d\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-ba1d5bb329d140eb8d8ca39bbbc9132d\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@5.2.0?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function maybeLoadScript(lib, version) {\n",
       "      var key = `${lib.replace(\"-\", \"\")}_version`;\n",
       "      return (VEGA_DEBUG[key] == version) ?\n",
       "        Promise.resolve(paths[lib]) :\n",
       "        new Promise(function(resolve, reject) {\n",
       "          var s = document.createElement('script');\n",
       "          document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "          s.async = true;\n",
       "          s.onload = () => {\n",
       "            VEGA_DEBUG[key] = version;\n",
       "            return resolve(paths[lib]);\n",
       "          };\n",
       "          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "          s.src = paths[lib];\n",
       "        });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else {\n",
       "      maybeLoadScript(\"vega\", \"5\")\n",
       "        .then(() => maybeLoadScript(\"vega-lite\", \"5.2.0\"))\n",
       "        .then(() => maybeLoadScript(\"vega-embed\", \"6\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}, \"axis\": {\"labelLimit\": 500}}, \"data\": {\"name\": \"data-c4fa829f441addc7c319d247141d7922\"}, \"mark\": {\"type\": \"rect\", \"stroke\": \"black\"}, \"encoding\": {\"color\": {\"field\": \"correlation\", \"scale\": {\"domain\": [0, 1]}, \"type\": \"quantitative\"}, \"facet\": {\"columns\": 3, \"field\": \"selection_condition\", \"type\": \"nominal\"}, \"tooltip\": [{\"field\": \"selection_condition\", \"type\": \"nominal\"}, {\"field\": \"antibody_concentration\", \"format\": \".3f\", \"type\": \"quantitative\"}, {\"field\": \"replicate_1\", \"type\": \"nominal\"}, {\"field\": \"replicate_2\", \"type\": \"nominal\"}, {\"field\": \"correlation\", \"format\": \".3f\", \"type\": \"quantitative\"}], \"x\": {\"field\": \"replicate_1\", \"title\": null, \"type\": \"nominal\"}, \"y\": {\"field\": \"replicate_2\", \"title\": null, \"type\": \"nominal\"}}, \"height\": {\"step\": 15}, \"width\": {\"step\": 15}, \"$schema\": \"https://vega.github.io/schema/vega-lite/v5.2.0.json\", \"datasets\": {\"data-c4fa829f441addc7c319d247141d7922\": [{\"selection_condition\": \"3e5vir_1200uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"0.0014_1\", \"replicate_2\": \"0.0014_1\", \"correlation\": 1.0}, {\"selection_condition\": \"3e5vir_1200uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"0.0014_2\", \"replicate_2\": \"0.0014_1\", \"correlation\": 0.3384577791831803}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"0.0014_1\", \"replicate_2\": \"0.0014_1\", \"correlation\": 1.0}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"0.0014_2\", \"replicate_2\": \"0.0014_1\", \"correlation\": 0.6207441478439452}, {\"selection_condition\": \"6e5vir_600uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"0.0014_1\", \"replicate_2\": \"0.0014_1\", \"correlation\": 1.0}, {\"selection_condition\": \"6e5vir_600uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"0.0014_2\", \"replicate_2\": \"0.0014_1\", \"correlation\": 0.8680051068823526}, {\"selection_condition\": \"3e5vir_1200uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"0.0014_1\", \"replicate_2\": \"0.0014_2\", \"correlation\": 0.3384577791831803}, {\"selection_condition\": \"3e5vir_1200uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"0.0014_2\", \"replicate_2\": \"0.0014_2\", \"correlation\": 1.0}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"0.0014_1\", \"replicate_2\": \"0.0014_2\", \"correlation\": 0.6207441478439452}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"0.0014_2\", \"replicate_2\": \"0.0014_2\", \"correlation\": 1.0}, {\"selection_condition\": \"6e5vir_600uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"0.0014_1\", \"replicate_2\": \"0.0014_2\", \"correlation\": 0.8680051068823526}, {\"selection_condition\": \"6e5vir_600uL\", \"antibody_concentration\": 0.0014, \"replicate_1\": \"0.0014_2\", \"replicate_2\": \"0.0014_2\", \"correlation\": 1.0}, {\"selection_condition\": \"3e5vir_1200uL\", \"antibody_concentration\": 0.0021, \"replicate_1\": \"0.0021_1\", \"replicate_2\": \"0.0021_1\", \"correlation\": 1.0}, {\"selection_condition\": \"3e5vir_1200uL\", \"antibody_concentration\": 0.0021, \"replicate_1\": \"0.0021_2\", \"replicate_2\": \"0.0021_1\", \"correlation\": 0.2947792594845062}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 0.0021, \"replicate_1\": \"0.0021_1\", \"replicate_2\": \"0.0021_1\", \"correlation\": 1.0}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 0.0021, \"replicate_1\": \"0.0021_2\", \"replicate_2\": \"0.0021_1\", \"correlation\": 0.5007008492510936}, {\"selection_condition\": \"6e5vir_600uL\", \"antibody_concentration\": 0.0021, \"replicate_1\": \"0.0021_1\", \"replicate_2\": \"0.0021_1\", \"correlation\": 1.0}, {\"selection_condition\": \"3e5vir_1200uL\", \"antibody_concentration\": 0.0021, \"replicate_1\": \"0.0021_1\", \"replicate_2\": \"0.0021_2\", \"correlation\": 0.2947792594845062}, {\"selection_condition\": \"3e5vir_1200uL\", \"antibody_concentration\": 0.0021, \"replicate_1\": \"0.0021_2\", \"replicate_2\": \"0.0021_2\", \"correlation\": 1.0}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 0.0021, \"replicate_1\": \"0.0021_1\", \"replicate_2\": \"0.0021_2\", \"correlation\": 0.5007008492510936}, {\"selection_condition\": \"6e5vir_1200uL\", \"antibody_concentration\": 0.0021, \"replicate_1\": \"0.0021_2\", \"replicate_2\": \"0.0021_2\", \"correlation\": 1.0}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.Chart(...)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check serum count correlations\n",
    "serum_df = get_corr_df(\n",
    "    'AUSAB-13', \n",
    "    [0.0014, 0.0021], \n",
    "    value_to_correlate = 'antibody_count'\n",
    ")[0]\n",
    "\n",
    "corr_heatmap(serum_df, \n",
    "             corr_col='correlation', \n",
    "             sample_cols=['replicate_1', 'replicate_2'], \n",
    "             group_col='selection_condition',\n",
    "#              corr_range=(0, 0.1)\n",
    "            )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "126256da",
   "metadata": {},
   "source": [
    "We see a similar scenario for serum selections. Note that correlation is slightly higher for the lower-potency selection, which seems consistent with generally seeing better correlation under conditions where the antibody is **less** neutralizing. Selection is more stochastic, vs conditions with less neutralization --> variant frequencies more dominated by library composition (nonvariable)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90336e22",
   "metadata": {},
   "source": [
    "## Variability in neut standard count between selection conditions\n",
    "\n",
    "We're seeing a huge drop in correlation between replicates when we convert raw variant counts to prob escape scores. Note that all selection samples were normalized to the **same** no-Antibody control. \n",
    "\n",
    "I made one large aliquot of virus mix, with libA and H6 spiked in at 1%, at a concentration of 6e5 virions / 300uL. I added 300uL of this mix for the 6e5-virion selections, 150uL for the 3e5-virion selections, and then infected 4 wells with 6e5 virions each for the no-Antibody controls. Each no-antibody well was prepped as an individual sample, and I then pooled 3 of these samples together as a single no-antibody control, to mimic the level of coverage I have for a typical experiment.\n",
    "\n",
    "The most likely explanation here is that correlations between raw counts are skewed artificially high, due to the uneven representation of variants in the library. Good correlation between high-count outlier variants is likely masking lower correlation for the bulk of the library. This gets factored out when we convert to prob_escape scores.\n",
    "\n",
    "It's still worth checking on proportion of neut standard counts for each selection, to see if this is affecting prob_escape scores."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "fa3138d2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>selection_condition</th>\n",
       "      <th>replicate</th>\n",
       "      <th>antibody_neut_standard_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3e5vir_1200uL</td>\n",
       "      <td>1</td>\n",
       "      <td>70125.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3e5vir_1200uL</td>\n",
       "      <td>2</td>\n",
       "      <td>130865.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6e5vir_1200uL</td>\n",
       "      <td>1</td>\n",
       "      <td>63749.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6e5vir_1200uL</td>\n",
       "      <td>2</td>\n",
       "      <td>55785.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6e5vir_600uL</td>\n",
       "      <td>1</td>\n",
       "      <td>9389.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6e5vir_600uL</td>\n",
       "      <td>2</td>\n",
       "      <td>10960.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  selection_condition replicate  antibody_neut_standard_count\n",
       "0       3e5vir_1200uL         1                       70125.0\n",
       "1       3e5vir_1200uL         2                      130865.0\n",
       "2       6e5vir_1200uL         1                       63749.0\n",
       "3       6e5vir_1200uL         2                       55785.0\n",
       "4        6e5vir_600uL         1                        9389.0\n",
       "5        6e5vir_600uL         2                       10960.0"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get full df of selection replicates for cocktail\n",
    "cocktail_df = get_corr_df('1C04-5G04', [1.37])[1]\n",
    "\n",
    "# get counts of antibody neut standard for each selection\n",
    "# very rough method of reducing large df to a summary of seleciton condition / rep / neut std count\n",
    "neut_std_cocktail = (cocktail_df.groupby(['selection_condition', 'replicate'])\n",
    "            .agg({'antibody_neut_standard_count': ['mean']})\n",
    "            .reset_index())\n",
    "neut_std_cocktail.columns = [''.join(col) for col in neut_std_cocktail.columns]\n",
    "neut_std_cocktail = neut_std_cocktail.rename(\n",
    "    columns={'antibody_neut_standard_countmean': 'antibody_neut_standard_count'}\n",
    ")\n",
    "\n",
    "neut_std_cocktail"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "64c1629c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>selection_condition</th>\n",
       "      <th>replicate</th>\n",
       "      <th>antibody_neut_standard_count</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>percent_neut_std</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3e5vir_1200uL</td>\n",
       "      <td>1</td>\n",
       "      <td>70125.0</td>\n",
       "      <td>2430772</td>\n",
       "      <td>2.884886</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3e5vir_1200uL</td>\n",
       "      <td>2</td>\n",
       "      <td>130865.0</td>\n",
       "      <td>2175773</td>\n",
       "      <td>6.014644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6e5vir_1200uL</td>\n",
       "      <td>1</td>\n",
       "      <td>63749.0</td>\n",
       "      <td>2976924</td>\n",
       "      <td>2.141439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6e5vir_1200uL</td>\n",
       "      <td>2</td>\n",
       "      <td>55785.0</td>\n",
       "      <td>2414912</td>\n",
       "      <td>2.310022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6e5vir_600uL</td>\n",
       "      <td>1</td>\n",
       "      <td>9389.0</td>\n",
       "      <td>2023872</td>\n",
       "      <td>0.463913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6e5vir_600uL</td>\n",
       "      <td>2</td>\n",
       "      <td>10960.0</td>\n",
       "      <td>2024530</td>\n",
       "      <td>0.541360</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  selection_condition replicate  antibody_neut_standard_count  total_counts  \\\n",
       "0       3e5vir_1200uL         1                       70125.0       2430772   \n",
       "1       3e5vir_1200uL         2                      130865.0       2175773   \n",
       "2       6e5vir_1200uL         1                       63749.0       2976924   \n",
       "3       6e5vir_1200uL         2                       55785.0       2414912   \n",
       "4        6e5vir_600uL         1                        9389.0       2023872   \n",
       "5        6e5vir_600uL         2                       10960.0       2024530   \n",
       "\n",
       "   percent_neut_std  \n",
       "0          2.884886  \n",
       "1          6.014644  \n",
       "2          2.141439  \n",
       "3          2.310022  \n",
       "4          0.463913  \n",
       "5          0.541360  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get total counts for each selection as well, to calculate percent neut standard\n",
    "\n",
    "sample_dict = {\n",
    "    1: '6e5vir_600uL_1',\n",
    "    2: '6e5vir_600uL_2',\n",
    "    3: '6e5vir_1200uL_1',\n",
    "    4: '6e5vir_1200uL_2',\n",
    "    5: '3e5vir_1200uL_1',\n",
    "    6: '3e5vir_1200uL_2'\n",
    "}\n",
    "\n",
    "total_counts = []\n",
    "selection_condition = []\n",
    "replicate = []\n",
    "\n",
    "# pull total counts as number valid barcode reads from barcode fates file\n",
    "for sample in sample_dict:\n",
    "    bc_fates = pd.read_csv(\n",
    "        f'results/barcode_runs/fates_by_sample/libB_221202_1_antibody_1C04-5G04_1.37_{sample}.csv'\n",
    "    )\n",
    "    \n",
    "    total_counts.append(bc_fates.loc[bc_fates['fate'] == 'valid barcode']['count'][3])\n",
    "    \n",
    "    selection_condition.append(\"_\".join(sample_dict[sample].split('_')[:2]))\n",
    "    \n",
    "    replicate.append(sample_dict[sample].split('_')[-1:][0])\n",
    "    \n",
    "total_counts_df = pd.DataFrame(\n",
    "    {'selection_condition': selection_condition,\n",
    "     'replicate': replicate,\n",
    "     'total_counts': total_counts}\n",
    ")\n",
    "\n",
    "cocktail_counts_summary = neut_std_cocktail.merge(\n",
    "    total_counts_df,\n",
    "    how='inner'\n",
    ")\n",
    "\n",
    "cocktail_counts_summary['percent_neut_std'] = (\n",
    "    cocktail_counts_summary['antibody_neut_standard_count'] / \n",
    "    cocktail_counts_summary['total_counts']) * 100\n",
    "\n",
    "cocktail_counts_summary"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2c45f78",
   "metadata": {},
   "source": [
    "Percent neut standard counts actually look quite consistent between replicates, with the exception of 3e5 virions. \n",
    "\n",
    "I noticed that the avg_prob_escape profiles look very strange for the 3e5-vir selections - very inconsistent trends, tend to spike up at the most potent Ab concentration (i.e. more escape). I wonder if this has to do with this neut standard variation between samples. Neut standard is at a consistent fraction, 1% of virions. But maybe I need to start with a minimum threshold of neut standard virions to get consistent growth across samples. Noise is going to be much more consequential when everything is starting at lower counts. So consider spiking up the neut standard if I do end up dropping virion number (maybe 2%)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34a6410c",
   "metadata": {},
   "source": [
    "## Main conclusions\n",
    "1. With technical replicates of selections, correlation between prob_escape scores for individual variants is very low across the board\n",
    "2. Reducing antibody saturation also leads to drop in prob_escape correlation\n",
    "    * At saturating conditions, correlations are likely skewed high by free virions. I.e. non-neutralized library will be very consistent between replicate infections, whereas neutralization is more stochastic. \n",
    "3. Neut standard counts are quite variable between samples for the 3e5-virion selections\n",
    "    * Would need to spike up neut standard fraction if I moved forward with these conditions\n",
    "    * As things stand, we can't really compare these replicates because the inconsistent neut standard counts are likely skewing results. Meaning that I can't make any conclusions about whether reducing virion number will introduce more severe bottlenecking. \n",
    "4. Increasing R1 primer concentration does not improve correlation in prob_escape scores between technical replicates."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  },
  "nbsphinx": {
   "orphan": true
  },
  "papermill": {
   "default_parameters": {},
   "duration": 13.55676,
   "end_time": "2022-12-02T20:42:26.186658",
   "environment_variables": {},
   "exception": null,
   "input_path": "dms-vep-pipeline/notebooks/fit_polyclonal.ipynb",
   "output_path": "results/notebooks/fit_polyclonal_libB_221202_1_1C04-5G04_1.ipynb",
   "parameters": {
    "n_threads": 2,
    "pickle_file": "results/polyclonal_fits/libB_221202_1_1C04-5G04_1.pickle",
    "prob_escape_csv": "results/prob_escape/libB_221202_1_1C04-5G04_1_prob_escape.csv"
   },
   "start_time": "2022-12-02T20:42:12.629898",
   "version": "2.3.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
