{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Single and multi-mutant primers for nearby epitopes\n",
    "\n",
    "The purpose of this notebook is to generate a list of primers that include random mutations at important epitope sites. For pairs of epitopes that are close to each other, double-mutant primers are also generated. The codons to mutate are from the previously defined 22 major epitope sites. These will just be NNK mutants, but I want to also take into account what the actual primer lengths will be based on Tm requirements."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqUtils import MeltingTemp as mt\n",
    "import Bio"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, manually define a list of the 22 major epitope sites identified in `data/ranked-list-of-epitope-positions-for-DMS-experiments.pdf`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# These epitopes are listed in HA1 numbering\n",
    "epitopes_ha1 = [193, 142, 135, 138, 160, 186, 159, 261, 121, 128, 131, 144, \n",
    "                198, 171, 190, 197, 62, 192, 83, 92, 311, 326]\n",
    "\n",
    "# convert to sequential numbering for this analysis\n",
    "epitopes = [i+16 for i in epitopes_ha1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import WT nt sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in just the first entry of the sequence file\n",
    "hk19_wt_nts = next(SeqIO.parse('./data/hk-45-2019.fasta', 'fasta'))\n",
    "ks17_wt_nts = next(SeqIO.parse('./data/ks-14-2017.fasta', 'fasta'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ReverseComplement(seq):\n",
    "    \"\"\"Returns reverse complement of sequence. Preserves upper/lower case.\"\"\"\n",
    "    d = {'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C', 'a': 't', 't': 'a', 'c': 'g',\n",
    "         'g': 'c', 'N':'N'}\n",
    "    rc = [d[nt] for nt in seq]\n",
    "    rc.reverse()\n",
    "    return ''.join(rc)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Single epitope primers\n",
    "Generate random `NNS` primers for each epitope, and write out to a csv. Note that IDT does not accept `S` as a random base, so I manually generate both `NNG` and `NNC` primers for each eptiope. This code is largely copied from `create_primers.py` in [TargetedTilingPrimers](https://github.com/jbloomlab/TargetedTilingPrimers/blob/main/create_primers.py)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def gen_single_primers(wt_nts, epitopes, name_prefix, random_codon_fwd_1, random_codon_fwd_2, outfile):\n",
    "    \n",
    "    # default primer settings\n",
    "    primerlength = 37\n",
    "    maxprimertm = 61\n",
    "    minprimertm = 60\n",
    "    minlength = 25\n",
    "    maxlength = 51\n",
    "    \n",
    "    wt_aa = wt_nts.seq.translate()\n",
    "    ncodons = len(wt_aa)\n",
    "    initial_flanklength = (primerlength - 3) // 2\n",
    "    \n",
    "    primers = []\n",
    "    for icodon in range(ncodons):\n",
    "        if icodon + 1 in epitopes:\n",
    "            i = icodon * 3\n",
    "            five_prime = wt_nts.seq[i - initial_flanklength : i]\n",
    "            three_prime = wt_nts.seq[i+3 : i+3+initial_flanklength]\n",
    "            primer = f'{str(five_prime)}{random_codon_fwd_1}{str(three_prime)}'\n",
    "            name = f'{name_prefix}-for-mut_epi-{icodon + 1}_{random_codon_fwd_1}'\n",
    "            primerseq = Bio.Seq.Seq(primer)\n",
    "            \n",
    "            TmNN = ('%0.2f' % mt.Tm_NN(primerseq, strict=False))\n",
    "            \n",
    "            add_3 = True\n",
    "            minus_5 = True\n",
    "            flank5 = flank3 = initial_flanklength\n",
    "            if float(TmNN) > float(maxprimertm):\n",
    "                while (float(TmNN) > float(maxprimertm) and\n",
    "                        len(primer) > minlength):\n",
    "                    if minus_5:\n",
    "                        flank5 -= 1\n",
    "                        five_prime = wt_nts.seq[i - (flank5): i]\n",
    "                        three_prime = wt_nts.seq[i + 3: i + 3 + flank3]\n",
    "                        primer = f\"{str(five_prime)}{random_codon_fwd_1}{str(three_prime)}\"\n",
    "                        minus_5 = False\n",
    "                    else:\n",
    "                        flank3 -= 1\n",
    "                        five_prime = wt_nts.seq[i - (flank5): i]\n",
    "                        three_prime = wt_nts.seq[i + 3: i + 3 + flank3]\n",
    "                        primer = f\"{str(five_prime)}{random_codon_fwd_1}{str(three_prime)}\"\n",
    "                        minus_5 = True\n",
    "                    primerseq = Bio.Seq.Seq(primer)\n",
    "                    TmNN = ('%0.2f' % mt.Tm_NN(primerseq, strict=False))\n",
    "                    primerlength = len(primer)\n",
    "                    \n",
    "            else:\n",
    "                if float(TmNN) < float(minprimertm):\n",
    "                    while (float(TmNN) < float(minprimertm) and\n",
    "                            len(primer) < maxlength):\n",
    "                        if add_3:\n",
    "                            flank3 += 1\n",
    "                            five_prime = wt_nts.seq[i - (flank5): i]\n",
    "                            three_prime = wt_nts.seq[i + 3: i + 3 + flank3]\n",
    "                            primer = f\"{str(five_prime)}{random_codon_fwd_1}{str(three_prime)}\"\n",
    "                            add_3 = False\n",
    "                        else:\n",
    "                            flank5 += 1\n",
    "                            five_prime = wt_nts.seq[i - (flank5): i]\n",
    "                            three_prime = wt_nts.seq[i + 3: i + 3 + flank3]\n",
    "                            primer = f\"{str(five_prime)}{random_codon_fwd_1}{str(three_prime)}\"\n",
    "                            add_3 = True\n",
    "                        primerseq = Bio.Seq.Seq(primer)\n",
    "                        TmNN = ('%0.2f' % mt.Tm_NN(primerseq,\n",
    "                                                   strict=False))\n",
    "                        primerlength = len(primer)\n",
    "                        \n",
    "                else:\n",
    "                    pass\n",
    "            \n",
    "            # sub in second random primer\n",
    "            primer2 = f\"{str(five_prime)}{random_codon_fwd_2}{str(three_prime)}\"\n",
    "            name2 = f'{name_prefix}-for-mut_epi-{icodon + 1}_{random_codon_fwd_2}'\n",
    "            \n",
    "            # also define reverse primer\n",
    "            rev_name1 = name.replace('for', 'rev')\n",
    "            rev_name2 = name2.replace('for', 'rev')\n",
    "            \n",
    "            # because the third base of the random codon is either G or C, we can just take the\n",
    "            # reverse compliment for the rev primer\n",
    "            rev_primer1 = ReverseComplement(primer)\n",
    "            rev_primer2 = ReverseComplement(primer2)\n",
    "            \n",
    "            primers.append((name, primer))\n",
    "            primers.append((name2, primer2))\n",
    "            primers.append((rev_name1, rev_primer1))\n",
    "            primers.append((rev_name2, rev_primer2))\n",
    "    \n",
    "    # sort primer list to get fwd and rev pools together\n",
    "    primers_sorted = sorted(primers, key = lambda x: x[0])\n",
    "    \n",
    "    f = open(outfile, 'w')\n",
    "    for (name, primer) in primers_sorted:\n",
    "        f.write(f'{name}, {primer}\\r\\n')\n",
    "        \n",
    "    return primers_sorted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generated 88 random primers for HK19 and 88 random primers for KS17.\n"
     ]
    }
   ],
   "source": [
    "hk19_primers = gen_single_primers(hk19_wt_nts, epitopes, 'hk19', 'NNG', 'NNC', \n",
    "                                  'results/hk19_single_epitope_primers.csv')\n",
    "\n",
    "ks17_primers = gen_single_primers(ks17_wt_nts, epitopes, 'ks17', 'NNG', 'NNC', \n",
    "                                  'results/ks17_single_epitope_primers.csv')\n",
    "\n",
    "print(f'Generated {len(hk19_primers)} random primers for HK19 and {len(ks17_primers)} random primers for KS17.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These numbers match up - fwd and reverse primers either with G or C in the third position for each of 22 epitopes."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Paired epitope primers"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There's definitely a more elegant way to do this, but for a one-shot analysis this is working. I loop through every defined epitope position, and generate a primer sequence that includes that epitope at the extreme 5' or 3' end of the primer. (I'm currently saying the primer would have 3 NT's between the terminus and the first mutation site, but I'm realizing this might be too short. I can edit this). Primer length is adjusted to fall in the target Tm. I then add the primer name and sequence to a list."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def gen_edge_primers(wt_nts, epitopes):\n",
    "    \n",
    "    # default primer settings\n",
    "    primerlength = 37\n",
    "    maxprimertm = 61\n",
    "    minprimertm = 60\n",
    "    minlength = 25\n",
    "    maxlength = 51\n",
    "    \n",
    "    wt_aa = wt_nts.seq.translate()\n",
    "    ncodons = len(wt_aa)\n",
    "    \n",
    "    # flank length is the number of nucleotides between the end of the primer and\n",
    "    # the first site to mutate\n",
    "    flank_length = 8\n",
    "    fiveprime_epitope_primers = []\n",
    "    threeprime_epitope_primers = []\n",
    "    for icodon in range(ncodons):\n",
    "        if icodon + 1 in epitopes:\n",
    "\n",
    "            # add a primer where the epitope is one codon away from the 5' end\n",
    "            five_prime_start = ((icodon+1) * 3) - 3 - flank_length # returns 0-indexed starting *nt* posn of primer\n",
    "            primer = wt_nts.seq[five_prime_start : five_prime_start + primerlength]\n",
    "\n",
    "            new_primerlength = primerlength\n",
    "            TmNN = ('%0.2f' % mt.Tm_NN(primer, strict=False))\n",
    "\n",
    "            # shorten primer if Tm is too high\n",
    "            if float(TmNN) > float(maxprimertm):\n",
    "                while (float(TmNN) > float(maxprimertm) and len(primer) > minlength):\n",
    "                    # only change length from 3' end\n",
    "                    new_primerlength -= 1\n",
    "                    primer = wt_nts.seq[five_prime_start : five_prime_start + new_primerlength]\n",
    "\n",
    "                    # update Tm\n",
    "                    TmNN = ('%0.2f' % mt.Tm_NN(primer, strict=False))\n",
    "            # lengthen primer if Tm is too low\n",
    "            elif float(TmNN) < float(minprimertm):\n",
    "                while (float(TmNN) < float(minprimertm) and len(primer) < maxlength):\n",
    "                    new_primerlength += 1\n",
    "                    primer = wt_nts.seq[five_prime_start : five_prime_start + new_primerlength]\n",
    "\n",
    "                    TmNN = ('%0.2f' % mt.Tm_NN(primer, strict=False))\n",
    "            else:\n",
    "                pass\n",
    "\n",
    "            fiveprime_epitope_primers.append((icodon+1, str(primer)))\n",
    "\n",
    "            # and repeat to add a primer where the epitope is one codon away from the 3' end\n",
    "\n",
    "            three_prime_end = ((icodon+1) * 3) + flank_length # returns 0-indexed ending *nt* posn of primer\n",
    "            primer = wt_nts.seq[three_prime_end - primerlength + 1 : three_prime_end + 1]\n",
    "\n",
    "            new_primerlength = primerlength\n",
    "            TmNN = ('%0.2f' % mt.Tm_NN(primer, strict=False))\n",
    "            if float(TmNN) > float(maxprimertm):\n",
    "                while (float(TmNN) > float(maxprimertm) and len(primer) > minlength):\n",
    "                    # only change length from 5' end\n",
    "                    new_primerlength -= 1\n",
    "                    primer = wt_nts.seq[three_prime_end - new_primerlength + 1 : three_prime_end + 1]\n",
    "\n",
    "                    TmNN = ('%0.2f' % mt.Tm_NN(primer, strict=False))\n",
    "            elif float(TmNN) < float(minprimertm):\n",
    "                while (float(TmNN) < float(minprimertm) and len(primer) < maxlength):\n",
    "                    new_primerlength += 1\n",
    "                    primer = wt_nts.seq[three_prime_end - new_primerlength + 1 : three_prime_end + 1]\n",
    "\n",
    "                    TmNN = ('%0.2f' % mt.Tm_NN(primer, strict=False))\n",
    "            else:\n",
    "                pass\n",
    "\n",
    "            threeprime_epitope_primers.append((icodon+1, str(primer)))\n",
    "    \n",
    "    primer_lists = [fiveprime_epitope_primers, threeprime_epitope_primers]\n",
    "    return primer_lists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "22 22\n",
      "22 22\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[(78, 'ttgatggagggaactgcacactaatagatgc'),\n",
       " (99, 'aaaataagaaatgggacctttttgttgaacgaagcag'),\n",
       " (108, 'aacgaagcagagcctacagcaactgtta'),\n",
       " (137, 'tggagtttaaaaatgaaagcttcaattgggctgg'),\n",
       " (144, 'tcaattgggctggagtcactcaaaacg'),\n",
       " (147, 'ctggagtcactcaaaacggaaaaagtttttcttg'),\n",
       " (151, 'aaaacggaaaaagtttttcttgcataaggggatctag'),\n",
       " (154, 'aaagtttttcttgcataaggggatctagtagtagtttctttagt'),\n",
       " (158, 'gcataaggggatctagtagtagtttctttagtagattaaattggt'),\n",
       " (160, 'ggggatctagtagtagtttctttagtagattaaattggttgacc'),\n",
       " (175, 'atttaaactacacatatccagcactgaacgtgactatg'),\n",
       " (176, 'taaactacacatatccagcactgaacgtgactatg'),\n",
       " (187, 'tgccaaacaaggaacaatttgacaaattgtacatttg'),\n",
       " (202, 'accacccgggtacggacaaggacca'),\n",
       " (206, 'cggacaaggaccaaatctccctgtatgc'),\n",
       " (208, 'aggaccaaatctccctgtatgctcaatcatca'),\n",
       " (209, 'accaaatctccctgtatgctcaatcatcaggaa'),\n",
       " (213, 'tgtatgctcaatcatcaggaagaatcacagtatctacc'),\n",
       " (214, 'atgctcaatcatcaggaagaatcacagtatctaccaa'),\n",
       " (277, 'tcaaaatacgaagtgggaaaagctcaataatgagatca'),\n",
       " (327, 'atgttaagcaaaacactctgaaattggcaacagg'),\n",
       " (342, 'taccagagaaacaaactagaggcatatttggcg')]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hk19_edge_primers = gen_edge_primers(hk19_wt_nts, epitopes)\n",
    "ks17_edge_primers = gen_edge_primers(ks17_wt_nts, epitopes)\n",
    "\n",
    "print(len(hk19_edge_primers[0]), len(hk19_edge_primers[1]))\n",
    "print(len(ks17_edge_primers[0]), len(ks17_edge_primers[1]))\n",
    "\n",
    "hk19_edge_primers[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We get two different output lists, each with 22 primers for each of the 22 epitopes. These entries are in the form `(epitope_site, primer_seq)`.\n",
    "\n",
    "Next, we can take the length for each primer, and ask whether any other epitopes are within that range. If there are, group them as a pair, and output an evenly centered primer within the length / Tm range specified above. The epitopes of interest will be replaced with `NNG` or `NNC`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def gen_paired_epitope_primers(wt_nts, edge_primers, epitopes, name_prefix, \n",
    "                               random_codon_fwd_1, random_codon_fwd_2, outfile):\n",
    "    \n",
    "    # default primer settings\n",
    "    primerlength = 37\n",
    "    maxprimertm = 61\n",
    "    minprimertm = 60\n",
    "    minlength = 25\n",
    "    maxlength = 51\n",
    "    \n",
    "    fiveprime_epitope_primers = edge_primers[0]\n",
    "    threeprime_epitope_primers = edge_primers[1]\n",
    "    \n",
    "    paired_epitopes = []\n",
    "    paired_epitope_primers = []\n",
    "    \n",
    "    flank_length = 8\n",
    "\n",
    "    for primer in fiveprime_epitope_primers:\n",
    "        # convert primer length into codon position range\n",
    "        epitope_1 = primer[0]\n",
    "        primer_seq = primer[1]\n",
    "        primer_nt_start = (epitope_1 * 3) - 2 - flank_length\n",
    "        primer_nt_end = primer_nt_start + len(primer_seq) - 1\n",
    "\n",
    "        # for each entry, query whether any of the other epitopes are in that range\n",
    "        for epitope_2 in epitopes:\n",
    "            if primer_nt_start < ((epitope_2*3)-2-flank_length) and primer_nt_end > ((epitope_2 * 3)+flank_length):\n",
    "                \n",
    "                # track which pairs we've already found\n",
    "                paired_epitopes.append((epitope_1, epitope_2)) \n",
    "                \n",
    "                # pull minimal sequence that's inclusive of the paired epitopes\n",
    "                epitopes_start = (epitope_1 * 3) - 2\n",
    "                epitopes_end = epitope_2 * 3\n",
    "                min_primer = wt_nts.seq[epitopes_start : epitopes_end+1]\n",
    "                TmNN = ('%0.2f' % mt.Tm_NN(min_primer, strict=False))\n",
    "\n",
    "                # lengthen primer on both sides until Tm is correct\n",
    "                minus_5 = True\n",
    "                flank5 = flank3 = flank_length\n",
    "                while (float(TmNN) < float(minprimertm) and len(primer) < maxlength):\n",
    "                    if minus_5:\n",
    "                        flank5 += 1\n",
    "                        paired_primer = wt_nts.seq[epitopes_start - flank5 : epitopes_end + flank3 + 1]\n",
    "                        primer_name = f'{name_prefix}-for-mut_epi-{epitope_1}-{epitope_2}'\n",
    "                        minus_5 = False\n",
    "                    else:\n",
    "                        flank3 += 1\n",
    "                        paired_primer = wt_nts.seq[epitopes_start - flank5 : epitopes_end + flank3 + 1]\n",
    "                        primer_name = f'{name_prefix}-for-mut_epi-{epitope_1}-{epitope_2}'\n",
    "                        minus_5 = True\n",
    "                    TmNN = ('%0.2f' % mt.Tm_NN(paired_primer, strict=False))\n",
    "\n",
    "                # change epitope sites to random codon by adding up seq strings based on flank lengths from previous step\n",
    "                # there's probably a better way to do this\n",
    "                fwd_paired_primer = str(paired_primer)\n",
    "                \n",
    "                primer_name1 = f'{name_prefix}-for-mut_epi-{epitope_1}-{epitope_2}_{random_codon_fwd_1}'\n",
    "                random_fwd_primer1 = fwd_paired_primer[:flank5-1]+random_codon_fwd_1+fwd_paired_primer[flank5+2:len(fwd_paired_primer)-flank3-4]+random_codon_fwd_1+fwd_paired_primer[len(fwd_paired_primer)-flank3-1:]\n",
    "                \n",
    "                primer_name2 = f'{name_prefix}-for-mut_epi-{epitope_1}-{epitope_2}_{random_codon_fwd_2}'\n",
    "                random_fwd_primer2 = fwd_paired_primer[:flank5-1]+random_codon_fwd_2+fwd_paired_primer[flank5+2:len(fwd_paired_primer)-flank3-4]+random_codon_fwd_2+fwd_paired_primer[len(fwd_paired_primer)-flank3-1:]\n",
    "                \n",
    "                primer_name3 = f'{name_prefix}-for-mut_epi-{epitope_1}-{epitope_2}_{random_codon_fwd_1}_{random_codon_fwd_2}'\n",
    "                random_fwd_primer3 = fwd_paired_primer[:flank5-1]+random_codon_fwd_1+fwd_paired_primer[flank5+2:len(fwd_paired_primer)-flank3-4]+random_codon_fwd_2+fwd_paired_primer[len(fwd_paired_primer)-flank3-1:]\n",
    "                \n",
    "                primer_name4 = f'{name_prefix}-for-mut_epi-{epitope_1}-{epitope_2}_{random_codon_fwd_2}_{random_codon_fwd_1}'\n",
    "                random_fwd_primer4 = fwd_paired_primer[:flank5-1]+random_codon_fwd_2+fwd_paired_primer[flank5+2:len(fwd_paired_primer)-flank3-4]+random_codon_fwd_1+fwd_paired_primer[len(fwd_paired_primer)-flank3-1:]\n",
    "                \n",
    "                fwd_random_primers = [(primer_name1, random_fwd_primer1), (primer_name2, random_fwd_primer2), \n",
    "                                      (primer_name3, random_fwd_primer3), (primer_name4, random_fwd_primer4)]\n",
    "                \n",
    "                # get reverse complement and add both to primer list\n",
    "                for fwd_primer in fwd_random_primers:\n",
    "                    rev_primer_name = fwd_primer[0].replace('for', 'rev')\n",
    "                    rev_primer = ReverseComplement(fwd_primer[1])\n",
    "                    paired_epitope_primers.append(fwd_primer)\n",
    "                    paired_epitope_primers.append((rev_primer_name, rev_primer))\n",
    "    \n",
    "    # now check whether there's any additional pairs in threeprime_epitope_primers\n",
    "    for primer in threeprime_epitope_primers:\n",
    "        # convert primer length into codon position range\n",
    "        epitope_2 = primer[0]\n",
    "        primer_seq = primer[1]\n",
    "        primer_nt_end = (epitope_2 * 3) + flank_length\n",
    "        primer_nt_start = primer_nt_end - len(primer_seq) + 1\n",
    "\n",
    "        # for each entry, query whether any of the other epitopes are in that range\n",
    "        for epitope_1 in epitopes:\n",
    "            if primer_nt_start < ((epitope_1*3)-2-flank_length) and primer_nt_end > ((epitope_1 * 3)+flank_length):\n",
    "                if (epitope_1, epitope_2) not in paired_epitopes:\n",
    "                    # Just throw an error if something's missing\n",
    "                    # everything should be covered by the fiveprime edge primers\n",
    "                    raise ValueError(f'missing pair: {epitope_1}_{epitope_2}')\n",
    "                    \n",
    "    # sort primer list to get fwd and rev pools together\n",
    "    paired_epitope_primers_sorted = sorted(paired_epitope_primers, key = lambda x: x[0])\n",
    "    \n",
    "    f = open(outfile, 'w')\n",
    "    for (name, primer) in paired_epitope_primers_sorted:\n",
    "        f.write(f'{name}, {primer}\\r\\n')\n",
    "    \n",
    "    return paired_epitope_primers_sorted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generated 80 paired primers for HK19 and 72 paired primers for KS17.\n"
     ]
    }
   ],
   "source": [
    "hk19_paired_primers = gen_paired_epitope_primers(hk19_wt_nts, hk19_edge_primers, \n",
    "                                                 epitopes, 'hk19', 'NNG', 'NNC',\n",
    "                                                 'results/hk19_paired_epitope_primers.csv')\n",
    "\n",
    "ks17_paired_primers = gen_paired_epitope_primers(ks17_wt_nts, ks17_edge_primers, \n",
    "                                                 epitopes, 'ks17', 'NNG', 'NNC', \n",
    "                                                 'results/ks17_paired_epitope_primers.csv')\n",
    "\n",
    "print(f'Generated {len(hk19_paired_primers)} paired primers for HK19 and {len(ks17_paired_primers)} paired primers for KS17.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "My method for adding the `NNG` / `NNC`s was a little weird, but I went through the sample HK19 sequence and verified that these primers actually line up with the correct sequence and epitope position."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Format for IDT oPools ordering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "hk19_primers = pd.read_csv('results/hk19_single_epitope_primers.csv', header=None)\n",
    "hk19_paired_primers = pd.read_csv('results/hk19_paired_epitope_primers.csv', header=None)\n",
    "ks17_primers = pd.read_csv('results/ks17_single_epitope_primers.csv', header=None)\n",
    "ks17_paired_primers = pd.read_csv('results/ks17_paired_epitope_primers.csv', header=None)\n",
    "\n",
    "hk19_primers.columns = ['Pool_name', 'Sequence']\n",
    "hk19_paired_primers.columns = ['Pool_name', 'Sequence']\n",
    "ks17_primers.columns = ['Pool_name', 'Sequence']\n",
    "ks17_paired_primers.columns = ['Pool_name', 'Sequence']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "hk19_opool = hk19_primers.replace(regex=[r'(?<=hk19-for).*', r'(?<=hk19-rev).*'], \n",
    "                                         value='-single-epi-pool')\n",
    "\n",
    "hk19_paired_opool = hk19_paired_primers.replace(regex=[r'(?<=hk19-for).*', r'(?<=hk19-rev).*'], \n",
    "                                                value='-paired-epi-pool')\n",
    "\n",
    "ks17_opool = ks17_primers.replace(regex=[r'(?<=ks17-for).*', r'(?<=ks17-rev).*'], \n",
    "                                         value='-single-epi-pool')\n",
    "\n",
    "ks17_paired_opool = ks17_paired_primers.replace(regex=[r'(?<=ks17-for).*', r'(?<=ks17-rev).*'], \n",
    "                                                value='-paired-epi-pool')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "opool_list = [(hk19_opool, 'hk19_opool'), (hk19_paired_opool, 'hk19_paired_opool'), \n",
    "              (ks17_opool, 'ks17_opool'), (ks17_paired_opool, 'ks17_paired_opool')]\n",
    "\n",
    "for pool, name in opool_list:\n",
    "    pool.to_csv(f'results/{name}.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "opools = pd.concat([hk19_opool, hk19_paired_opool, ks17_opool, ks17_paired_opool])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "from plotnine import *\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAi8AAAGuCAYAAABY0OakAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAAPYQAAD2EBqD+naQAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAgAElEQVR4nO3de3BU9f3/8dfZDRuSJTHZLkFaGgJWHe0g46XIRaCgUFuKtA2CYrVRAbWl2sZawcHGfLFDZ4rWlkix2Cmto8DY4ExFHbx0kBLUar0MUvBCSaiXATcbDC65kN3z+8NftgSScHbJ4exn83zMdOqe/eTDe8/7nJPXnrM5a9m2bQsAAMAQPq8LAAAASAXhBQAAGIXwAgAAjEJ4AQAARiG8AAAAoxBeAACAUQgvAADAKIQXAABglByvC3BDJBJxbe5AIKD29nbX5veaZVnKy8tTS0uLsvX+hfTQfPTQfPTQfG70MBwOOxrHmZcU5ebmel2Cq3w+n/Lz8+XzZe+mQQ/NRw/NRw/N52UPs3etAgCArER4AQAARiG8AAAAoxBeAACAUQgvAADAKJadhX/D1dzc7NqnoHNyctTR0eHK3JnAsqzkn79l4aYhiR5mA3poPnpoPjd66PR3d1be56W9vd21+wcUFBTo0KFDrsydCfx+vwKBgGKxmOLxuNfluIIemo8emo8ems+NHjoNL1w2AgAARiG8AAAAoxBeAACAUQgvAADAKIQXAABgFMILAAAwCuEFAAAYJSvv8wIAmcayLFmW5XgsgJ4RXgDAZZZlKRQKOQ4l2XpHVqCvEF4AwGWdZ12qqqoUjUZ7HRsKhVRdXX2KKgPMRHgBgFMkGo2qsbHR6zIA4/GBXQAAYBTCCwAAMArhBQAAGIXwAgAAjEJ4AQAARiG8AAAAoxBeAACAUQgvAADAKIQXAABgFMILAAAwCuEFAAAYhfACAACMQngBAABGIbwAAACjEF4AAIBRCC8AAMAohBcAAGAUwgsAADAK4QUAABiF8AIAAIxCeAEAAEYhvAAAAKMQXgAAgFEILwAAwCiEFwAAYBTCCwAAMArhBQAAGIXwAgAAjEJ4AQAARiG8AAAAoxBeAACAUQgvAADAKDleFyBJNTU1eu2119TS0qKCggJNnz5dc+bMkSQ1NDRo5cqVqq+v15AhQ7Rw4UKNHj3a44oBAIBXMuLMyxVXXKGHHnpIGzZs0PLly/Xiiy9q27Zt6ujo0LJlyzRmzBitW7dOV111lZYvX66DBw96XTIAAPBIRoSX0tJS5ebmJh9blqWPPvpIO3bsUFtbm2bPnq0BAwZo4sSJKi0tVV1dnYfVAgAAL2XEZSNJ+vOf/6xNmzapra1NJSUlmjJlirZv366ysjL5fP/LWCNHjlRDQ4OHlQIAAC9lTHj5wQ9+oOuuu07vv/++Xn75ZQWDQbW0tCgYDHYZFwwGdeDAgS7LIpGIIpFI8rHP59PgwYNdqdOyLPn9flfmzgSdry2bXyM9NJ9pPbQsK62fM+k1psq0HqaK/dBdGRNepM9XxJlnnql//etfWrduncLhsGKxWJcxsVhMeXl5XZbV1tZqzZo1yccVFRVatGiRa3UGAgHX5s4UhYWFXpfgKnpoPnpoPnpoPq96mFHhpVMikdDHH3+sCy64QLW1tUokEslLR3v37tWkSZO6jC8vL9fkyZOTj30+n5qamlypLRgMHheosonf71dhYaGam5sVj8e9LscV9NB8pvXQsiwVFRWl/HP00Fzsh+kpLi52NM7z8PLZZ5/p1Vdf1cUXX6yBAwdq9+7deuaZZzR37lyNGjVKgUBAGzdu1KxZs/TKK6+ooaFBEyZM6DJHOBxWOBxOPo5EIq5tLLZtZ+2GeLR4PJ61r5Mems+0Hh79ub1U0EPz0UN3eB5eJOn555/XH/7wByUSCYVCIX3nO9/RjBkzZFmWli5dqpqaGq1fv14lJSVasmRJWu9gAABAdvA8vAwaNEi//OUve3y+rKxMK1asOIUVAQCATJYR93kBAABwivACAACMQngBAABGIbwAAACjEF4AAIBRCC8AAMAohBcAAGAUwgsAADAK4QUAABiF8AIAAIxCeAEAAEYhvAAAAKMQXgAAgFEILwAAwCiEFwAAYBTCCwAAMArhBQAAGIXwAgAAjEJ4AQAARiG8AAAAoxBeAACAUQgvAADAKIQXAABgFMILAAAwCuEFAAAYhfACAACMQngBAABGIbwAAACjEF4AAIBRCC8AAMAohBcAAGAUwgsAADAK4QUAABiF8AIAAIxCeAEAAEYhvAAAAKMQXgAAgFEILwAAwCiEFwAAYBTCCwAAMArhBQAAGIXwAgAAjEJ4AQAARiG8AAAAo+R4XYAbAoGAcnNzXZk7JydHBQUFrsydCSzLkiQFg0HZtu1xNe6gh+bL9h52oofmYj90+d/25F91WXt7u9rb212Zu6CgQIcOHXJl7kzg9/sVCAQUi8UUj8e9LscV9NB8pvXQ5/Ol9YaKHpqL/TA9TvcTLhsBAACjEF4AAIBRCC8AAMAohBcAAGAUwgsAADAK4QUAABiF8AIAAIxCeAEAAEYhvAAAAKMQXgAAgFEILwAAwCiEFwAAYBTCCwAAMArhBQAAGIXwAgAAjEJ4AQAARiG8AAAAoxBeAACAUQgvAADAKIQXAABgFMILAAAwCuEFAAAYhfACAACMQngBAABGIbwAAACjEF4AAIBRCC8AAMAoOV4XAKBvWZaV/H+fr/f3J7Zty7btlObunP9EUpk7lXk7nei1pSPV9QHAG4QXIItYlqXTTjtNklRUVHTC8bZtKxqNOvqFbVmWQqFQSuHFydzpzGtZlnJzc1Ma73Ss0/UBwDuEFyCLdJ7BqKqqUjQa7XVsKBRSdXW1LMtyHF7cmDuVeUeMGKHKykpHY1Mdn+r6AOAdwguQhaLRqBobG42a28m8xcXFKdWQ6ngAZuADuwAAwCiEFwAAYBTCCwAAMArhBQAAGIXwAgAAjEJ4AQAARiG8AAAAoxBeAACAUQgvAADAKIQXAABgFMILAAAwCuEFAAAYxfMvZjxy5IhWr16tt956S4cOHVI4HNacOXM0efJkSVJDQ4NWrlyp+vp6DRkyRAsXLtTo0aM9rhoAAHjF8zMv8XhcoVBI9957r9avX68f/ehH+v3vf6/du3ero6NDy5Yt05gxY7Ru3TpdddVVWr58uQ4ePOh12QAAwCOeh5eBAwfqmmuu0emnny7LsnTuuefqnHPO0a5du7Rjxw61tbVp9uzZGjBggCZOnKjS0lLV1dV5XTYAAPCI5+HlWK2trXr//fc1fPhw7du3T2VlZfL5/lfmyJEj1dDQ4GGFAADAS55/5uVoiURCDzzwgM4880ydf/75evfddxUMBruMCQaDOnDgQJdlkUhEkUgk+djn82nw4MGu1GhZlvx+vytzZ4LO15bNrzGbe2hZVso/4/P5HP2cW3OnM6+bnK6PVKQ7X7Zup1J274cSx1K3ZUx4sW1bq1atUjQaVXV1tSzLUl5enmKxWJdxsVhMeXl5XZbV1tZqzZo1yccVFRVatGiRa7UGAgHX5s4UhYWFXpfgqv7QQ6eKioqMnNstmVQz+6H56KE7MiK82Lat1atXa+/evVq2bFkynJSWlqq2tlaJRCJ56Wjv3r2aNGlSl58vLy9P/nWS9Pk7p6amJldqDQaDxwWqbOL3+1VYWKjm5mbF43Gvy3FFNvfQsqyUf/kePHhQtm17Nnc687rJ6fpIRbqvkf3QXBxL01NcXOxoXEaEl4ceekjvvPOO7r33XuXn5yeXjxo1SoFAQBs3btSsWbP0yiuvqKGhQRMmTOjy8+FwWOFwOPk4Eom4trHYtp21G+LR4vF41r7ObO7h0Z8PcyqRSCiRSHg2dzrzusnp+khFuq+R/dB89NAdnoeXAwcO6Omnn9aAAQN0ww03JJfPnj1bc+bM0dKlS1VTU6P169erpKRES5Ysyah3aQAA4NTyPLyUlJTob3/7W4/Pl5WVacWKFaewIgAAkMky63wtAADACRBeAACAUQgvAADAKGmFl6lTp2r37t3dPvfuu+9q6tSpJ1UUAABAT9IKL1u2bFFzc3O3zzU3N2vr1q0nVRQAAEBP0r5s1NPtrrdv366SkpK0CwIAAOiN4z+VXr58uZYvXy7p8+AyZcqU42681NbWpo6ODv3whz/s2yoBAAD+P8fhZfz48br99ttl27b+7//+T1dffbWGDRvWZUwgENA555yjmTNn9nmhAAAAUgrhZfLkycnvD7IsSwsWLNAXv/hF1woDAADoTlp32K2qqurrOgAAABxJK7wkEgk9/PDD+utf/6oPPvhAra2tXZ63LEt79uzpkwIBAACOllZ4ufPOO3Xfffdp8uTJmjJligKBQF/XBQAA0K20wsujjz6q6upq3X333X1dDwAAQK/SCi+tra0aP358X9fS71mW1eP9c45l27Zs23a5IvQHx97y4GTHpfoz6cwLoH9LK7xcc801evLJJ3XppZf2dT39lmVZCoVCKYWXaDRKgEHagsGgbNtWUVGRUXMDQFrhZezYsVq6dKn279+vadOmdXuA+t73vnfSxfUnnWddqqqqFI1Gex0bCoVUXV0ty7IIL0hbIBBwvM1J0ogRI1RZWdnnc6cyLwBIaYaXa6+9VpLU0NCgDRs2HPe8ZVmKx+MnV1k/FY1G1djY6HUZ6EecbnPFxcWuzJ3OvAD6t7TCy969e/u6DgAAAEfSCi/Dhw/v6zoAAAAcSSu87Nu374RjSktL05kaAACgV2mFl7KyshP+VQyfeQEAAG5IK7w88cQTxy1ramrS5s2b9fLLL+tXv/rVSRcGAADQnbTCy6xZs7pdXlFRocrKSr344ouaO3fuSRUGAADQnT6/teW3vvUtrV+/vq+nBQAAkORCeNm+fbsGDhzY19MCAABISvOy0a233nrcsvb2du3atUvbtm3Tz372s5MuDAAAoDtphZcnn3zyuGUDBw7UsGHDtGrVKs2fP/+kCwMAAOgOd9gFAABG4bvoAQCAUdIOL2+88YauvPJKDR06VLm5uRo6dKjmzJmjN998sy/rAwAA6CKty0b/+Mc/NG3aNJ1++um6+uqrNWTIEO3fv19PPPGExo0bp+eee06XXHJJX9cKAACQXnhZvHixvv71r2vTpk3KyfnfFL/+9a81Y8YMLV68WNu2beuzIgEAADqlddnojTfe0K233toluEiS3+/Xrbfeqtdff71PigMAADhWWmdegsGgDhw40O1z+/fvVzAYPKmigP7AsqwTfsFpJ9u2Zdu2yxVBknw+Z+/p6AngnbTCy8yZM3XnnXdq2LBhuuyyy5LLn3/+eS1ZskRXXHFFnxUIZCPLshQKhVIKL9FolF+WLgoGg7JtW0VFRY7G0xPAO2mFl/vuu087d+7UN77xDRUWFqqkpEQHDhxQc3Ozvva1r2nFihV9XSeQVTrPulRVVSkajfY6NhQKqbq6WpZl8YvSRYFAgJ4AhkgrvBQXF+ull17Spk2btG3bNjU1NSkUCumSSy7RjBkzHJ92Bfq7aDSqxsZGr8vAUegJkPnSCi8vvPCC9u3bp+uvv/64S0Rr167V8OHDNWXKlD4pEAAA4GhpnSJZunSp9u/f3+1zn3zyiZYuXXpSRQEAAPQkrfCyc+dOXXTRRd0+d8EFF2jnzp0nVRQAAEBP0govlmXp008/7fa5pqYmxePxkyoKAACgJ2mFl4svvlgPPvjgcZ+yt21bq1at0sUXX9wnxQEAABwrrQ/sVldXa8qUKTrvvPNUUVGhoUOH6qOPPtJf/vIXvfvuu9qyZUsflwkAAPC5tMLLuHHj9MILL+jnP/+57rzzTiUSCfl8vuTysWPH9nWdAAAAktIML5I0YcIE1dXVqaWlRU1NTSoqKlJ+fn5f1pa2QCCg3NxcV+bOyclRQUGBK3Onyo2vYei842vn3UazUSb1MBV87Ubmcbsn7Ifm4ljq8r99shPk5eUpLy+vL2rpM+3t7Wpvb3dl7oKCAh06dKjP5/X5fCkHrlgspkQi0ad1+P1+BQIBxWKxrP3gtVs9TIVb/U5nXqTP6T6Ybl/YD83FsTQ9TvcTboULAACMQngBAABGIbwAAACjEF4AAIBRCC8AAMAohBcAAGAUwgsAADAK4QUAABiF8AIAAIxCeAEAAEYhvAAAAKMQXgAAgFEILwAAwCiEFwAAYBTCCwAAMArhBQAAGIXwAgAAjEJ4AQAARiG8AAAAoxBeAACAUQgvAADAKIQXAABgFMILAAAwCuEFAAAYhfACAACMQngBAABGIbwAAACj5HhdAACYyudz9v7P6TgAzhBeACBFwWBQtm2rqKjI61KAfonwAgApCgQCsixLVVVVikajJxw/YsQIVVZWnoLKgP6B8AIAaYpGo2psbDzhuOLi4lNQDdB/cCEWAAAYhfACAACMQngBAABGIbwAAACjEF4AAIBRCC8AAMAohBcAAGAUwgsAADAK4QUAABiF8AIAAIxCeAEAAEbx/LuNNm3apL///e+qr6/XuHHjdMcddySfa2ho0MqVK1VfX68hQ4Zo4cKFGj16tIfVAgAAr3l+5iUUCmnOnDmaPn16l+UdHR1atmyZxowZo3Xr1umqq67S8uXLdfDgQY8qBQAAmcDz8DJ+/HiNHTtWhYWFXZbv2LFDbW1tmj17tgYMGKCJEyeqtLRUdXV1HlUKAAAygefhpSf79u1TWVmZfL7/lThy5Eg1NDR4WBUAAPCa55956UlLS4uCwWCXZcFgUAcOHDhubCQSUSQSST72+XwaPHiwK3VZliW/3+/KvKny+Xxp/VxvOl+b09eY6r9v23bKNfU1t3qYag2p8vv9XcJ8X82LzOX1duqmTNgP3ZTqsdREXvYwY8NLXl6eYrFYl2WxWEx5eXnHja2trdWaNWuSjysqKrRo0SLXagsEAq7NnYqioiLX5j72Ml5PbNt2/AszlbFuy5QeOhEMBmXbtk477TSvS8Ep5nQ/NJVJ+2G66KE7Mja8lJaWqra2VolEIvluc+/evZo0adJxY8vLyzV58uTkY5/Pp6amJlfqCgaDx4WqvmBZVsph5ODBg31+JsPv96uwsFDNzc2Kx+O9ju2suaqqStFotNexoVBI1dXVrtScKrd6mIpU+h0IBGRZlqP1PGLECFVWVvZFicgATvZDU2XCfuimVI6lpnKjh8XFxY7GeR5e4vG44vG4EomEEomE2tvb5fP5NGrUKAUCAW3cuFGzZs3SK6+8ooaGBk2YMOG4OcLhsMLhcPJxJBJxbWOxbduVuU90OaA7nevMDZ196U1nzdFoVI2NjY7mdbNmp9zqYSrS6beT9ex0x4cZnOyHpsqE/fBUoIfu8Dy8bNiwQevXr08+rqur09SpU/WTn/xES5cuVU1NjdavX6+SkhItWbLE1UslAAAg83keXubNm6d58+Z1+1xZWZlWrFhxiisCAACZLGP/VBoAAKA7hBcAAGAUwgsAADAK4QUAABiF8AIAAIxCeAEAAEYhvAAAAKMQXgAAgFEILwAAwCiEFwAAYBTCCwAAMArhBQAAGIXwAgAAjEJ4AQAARsnxugCkz+dzlj1t25Zt2y5XAwDAqUF4MVAwGJRt2yoqKnI03rZtRaNRAgwAICsQXgwUCARkWZaqqqoUjUZ7HRsKhVRdXS3LsggvAICsQHgxWDQaVWNjo9dlAABwSvGBXQAAYBTCCwAAMArhBQAAGIXwAgAAjEJ4AQAARiG8AAAAoxBeAACAUQgvAADAKIQXAABgFMILAAAwCuEFAAAYhfACAACMQngBAABGIbwAAACjEF4AAIBRCC8AAMAoOV4XAHjB53OW223blm3bLlcDpM+yLFmW5Wgs2zOyBeEF/YplWQoEAsrNzXU03rZtRaNRDvjISJZlKRQKpRRe2J6RDQgv6Fc636VWVVUpGo32OjYUCqm6ulqWZXGwR0Zie0Z/RXhBvxSNRtXY2Oh1GUCfYHtGf8MHdgEAgFEILwAAwCiEFwAAYBTCCwAAMArhBQAAGIXwAgAAjEJ4AQAARsnK+7ykcgfVVOXk5KigoMCVud0UDAYdjeu8U2cwGHTlRlZO68gkJtaM7ODGfpgp27Opx1Kn3D6WZgIve5iV4aW9vV3t7e2uzF1QUKBDhw71+bw+n8+1wCVJsVhMiUTihOP8fr8CgYBisZji8XivY9Op2WkdbsmUmt3uN7LDifbDTNme0+HWsTRTpHIsNZUbPXS6PXPZCAAAGIXwAgAAjEJ4AQAARiG8AAAAoxBeAACAUQgvAADAKIQXAABglKy8z4vbfD5nmc+27Yy5OZHTmjtvrOQ1y7Ic1+L2ena67ryeE9nHsqxet5X+sB2lciyQMuu4a5pUj7teIrykwLKslO7ea9u2otGop03uvLtjUVGRo/Feb5DS5+s5FAqltBO5sZ7TWXeZEv5grs7tzrIsx9tetkr1WCBlxnHXROkcd926GawThJcUdKbSqqoqRaPRXseGQiFVV1fLsixPd6JAIJByzV7LlPWcyrobMWKEKisrHY09ejxwrHS2u2yVyrFAypzjronSOe56ifCShmg0qsbGRq/LSAk1u1tHcXGx47FHjwd6ksp2l+0y5VjQH5iyrrP/gikAAMgqhBcAAGAUwgsAADAK4QUAABiF8AIAAIxCeAEAAEYhvAAAAKMQXgAAgFEILwAAwCiEFwAAYBTCCwAAMArhBQAAGIXwAgAAjEJ4AQAARiG8AAAAo+R4XUC28/mc5UOn49A91jPgTCbsA7Ztuzq/09do27brtThhWZYsy3I0NlNq9hrhxSXBYFC2bauoqMjrUrIa6xlwJp19xbbtlH6ppjK2vb3dcR1OpfoabdtWNBr1NAxYlqVQKJTSuvO65kxAeHFJIBCQZVmqqqpSNBo94fgRI0aosrLyFFSWXVjPgDPp7itOxqcyNhQKqbq6OqXanUrlNXbWYVmW5+HFtJozAeHFZdFoVI2NjSccV1xcfAqqyV6sZ8CZVPcVJ+NTGXsqZEodqTCxZi95f/ETAAAgBYQXAABgFMILAAAwCuEFAAAYhfACAACMQngBAABGIbwAAACjEF4AAIBRCC8AAMAohBcAAGAUwgsAADAK4QUAABjFiC9m/Oyzz/Tggw/q9ddfV15enr773e9q1qxZXpcFAAA8YER4eeihh3TkyBH96U9/0oEDB3T33Xdr2LBhuvDCC70uDQAAnGIZf9motbVVdXV1uvbaa5Wfn6+ysjJNnz5dzz33nNelAQAAD2R8ePnwww9l27aGDx+eXDZixAjt27fPw6oAAIBXMv6yUWtrq/Lz87ssCwaDamlpST6ORCKKRCLJxz6fT4MHD+7zWizLkiSFQqETji0qKnI8NtXxbo09ekxOTk7y9fYklfXROcbv98vn6z0zZ/t6zpQ6qLl/1ZEpNXeOsSxLOTm9/wpK5ViQbh1OjkmSZNv2CY+JR+uc80THUreOo1JqNadTh2VZ8vv9jubva5Zt27Yn/7JDe/bs0R133KGNGzcml9XV1enRRx/VqlWrJH3+mZg1a9Ykn6+oqNCiRYtcqSeVjSHVjd2tuamZOvpibKbUYWLNmVIHNZ/c3G7JlJozpQ4nMv7My5e+9CVJ0r59+1RaWipJ2rt3b/K/Jam8vFyTJ09OPvb5fGpqanKlnmAwqMOHDzsaa1mWUsmGqYx3a6zf71dBQYEOHTqkeDzu6Ge8lup6zs/Pd6WHmdLvnJwcDRo0yFEPM6XmVMfm5eV53kM3151bPcyUfkup7YepSLUOp9I581JYWOjpsdTtgJGfn69YLNancxYXFzsal/HhZeDAgZowYYIeeeQR/fSnP9Unn3yiZ599VrfddltyTDgcVjgcTj6ORCKubSy2baujo8OVuTNB507f0dFhTHhJFT0038CBA+mh4bJ9P+y8nJLtPfTqtWV8eJGkm266STU1NaqoqFBeXp7Ky8v5M2kAAPopI8LLoEGDtHjxYq/LAAAAGSDj/1QaAADgaIQXAABgFMILAAAwCuEFAAAYhfACAACMQngBAABGIbwAAACjEF4AAIBRCC8AAMAohBcAAGAUwgsAADAK4QUAABjFsju/ex2QFIlEVFtbq/LycoXDYa/LQRroofnoofnoobs484IuIpGI1qxZo0gk4nUpSBM9NB89NB89dBfhBQAAGIXwAgAAjOK/55577vG6CGSWvLw8XXTRRcrPz/e6FKSJHpqPHpqPHrqHD+wCAACjcNkIAAAYhfACAACMkuN1AfDGAw88oK1btyon53+bwIMPPqjBgwdLkt577z2tWbNGDQ0NCoVCuv766zVmzBivykUvtm/frscee0z79+9XYWGhbrzxRo0fP14NDQ1auXKl6uvrNWTIEC1cuFCjR4/2ulx0o6ce1tTUaOfOnfroo490880365vf/KbXpaIH3fVw+PDhWrt2rXbv3q2Ojg6VlZXphhtu0Jlnnul1ueaz0S/95je/sdeuXdvtc4cOHbK///3v288//7zd0dFhv/baa3Z5ebn94YcfnuIqcSJvvvmmff3119s7d+604/G43dTUZH/88cf2kSNH7BtvvNHesGGD3d7ebm/dutWeO3eu3dTU5HXJOEZPPbRt2960aZP95ptv2rfffrv99NNPe1wpetJTD9955x37mWeesT/99FO7o6PDfvLJJ+1rrrnGbmlp8bpk43HZCMfZtWuXgsGgLr30Uvn9fl144YU6++yztWXLFq9LwzEee+wxzZ07V+eee658Pp+Kiop0+umna8eOHWpra9Ps2bM1YMAATZw4UaWlpaqrq/O6ZByjpx5K0owZMzR69GgFAgGPq0RveurhWWedpcsvv1yFhYXy+/369re/rdbWVn3wwQdel2w8Lhv1Y5s3b9bmzZsVDoc1c+ZMTZs2Lfmcfcwfodm2rfr6+lNcIXoTj8f13nvvacyYMbr55pvV2tqq888/X/Pnz9e+fftUVlYmn+9/709GjhyphoYGDyvGsXrrYTAY9Lo8OJBKD9977z3Ztq2hQ4d6VG324MxLPzVz5kytXr1ajzzyiObPn6+1a9dq+/btkqSzzz5bzc3NevbZZ9XR0aFXX31Vu3btUltbm8dV42gHDx5UR0eHtm7dqnvvvVc1NTU6eCmNJMAAAAP3SURBVPCgHn74YbW0tBx34AwGg2ppafGoWnSntx7CDE572NzcrPvvv1/z5s0jmPYBwks/dcYZZyRPZZ533nmaMWNG8pJCYWGh7r77bj377LO67rrr9Mwzz+iSSy7RF77wBY+rxtFyc3MlfX5pIRwOa9CgQbryyiv16quvKi8vT7FYrMv4WCymvLw8L0pFD3rrIczgpIexWEz33HOPLrjgApWXl3tValbhshEkSZZldblUdO6552rFihXJx3fccYcuu+wyL0pDDwYNGqRwOCzLso57rrS0VLW1tUokEslLR3v37tWkSZNOdZnoRW89hBlO1MNYLKaqqip95Stf0YIFC05xddmLMy/91LZt23T48GElEgn9+9//1lNPPaWxY8cmn9+zZ4+OHDmi1tZWPf744/r00081depUDytGd6ZPn66nnnpKTU1NOnz4sGprazVmzBiNGjVKgUBAGzdu1JEjR7Rt2zY1NDRowoQJXpeMY/TUQ0k6cuSI2tvblUgkFI/H1d7erng87nHFOFZPPTx8+LDuueceffnLX9Ytt9zidZlZha8H6KcWL16shoYGJRKJ5Ad2L7/88uTz999/v/75z3/Ktm2NGjVK8+fPT/4FBDJHPB7XH//4R23ZskV+v18XXXSRFixYoPz8fNXX16umpkb19fUqKSnRTTfdxH1eMlBvPbzrrrv09ttvdxl/22236dJLL/WoWnSnpx6+9NJL+u1vf6vc3NwuZ2aqqqr01a9+1cOKzUd4AQAARuGyEQAAMArhBQAAGIXwAgAAjEJ4AQAARiG8AAAAoxBeAACAUQgvAADAKIQXAABgFMILAAAwCuEFgKssyzrh/9auXet1mQAMwrdKA3DVSy+91OXxuHHj9OMf/1jz5s1LLjvjjDNOdVkADEZ4AeCqo7+tvFNpaWm3ywHACS4bAcgIW7ZskWVZ2rx5s+bMmaNBgwaptLRUjz32mCTpd7/7nUpLSxUKhTR//ny1tbV5XDEAr3DmBUBGueWWW1RRUaEFCxZozZo1uvbaa/XWW2/p7bff1urVq/Wf//xHlZWVGjlypO666y6vywXgAcILgIxy5ZVX6he/+IUkacyYMdq4caPWrVunPXv2aMCAAZI+P0vz+OOPE16AforLRgAyyrRp05L/fdppp6mkpESTJk1KBhdJOuuss/Tf//7Xi/IAZADCC4CMUlRU1OVxIBDodllra+upLAtABiG8AAAAoxBeAACAUQgvAADAKIQXAABgFMu2bdvrIgAAAJzizAsAADAK4QUAABiF8AIAAIxCeAEAAEYhvAAAAKMQXgAAgFEILwAAwCiEFwAAYBTCCwAAMArhBQAAGIXwAgAAjPL/AFdW8hF8xXMWAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<bound method ggplot.draw of <ggplot: (8790218000597)>>"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "opools['Tm'] = opools.apply(lambda x: '%0.2f' % mt.Tm_NN(x.Sequence, strict=False), axis=1)\n",
    "opools['Tm'] = opools['Tm'].astype('float')\n",
    "\n",
    "p = (ggplot(opools) + \n",
    "     aes('Tm') + \n",
    "     geom_histogram(color='white', bins=40)\n",
    "    )\n",
    "\n",
    "p.draw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
